{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a68179f1-6218-4af0-b352-79dccdb72e82",
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_file_path = 'data/RESPOND_247_coding_final.maf'\n",
    "expression_file_path = 'data/Expression_remove_BE.txt'\n",
    "target_gene='RPL10'\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statistics import mean \n",
    "from scipy.stats import ranksums\n",
    "import matplotlib.pylab as plt\n",
    "from statsmodels.stats import multitest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12bdca97-76e8-4627-8c49-51098cc5fd8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_maf_data(file_path):\n",
    "    # Define columns of interest\n",
    "    columns = [\"Hugo_Symbol\", \"Tumor_Sample_Barcode\"]\n",
    "    df = pd.read_csv(file_path, sep='\\t', comment=\"#\", usecols=columns)\n",
    "    df.rename(columns={'Hugo_Symbol': 'gene', 'Tumor_Sample_Barcode': 'sample'}, inplace=True)\n",
    "    df['mutation'] = 1\n",
    "    return df\n",
    "    \n",
    "maf_df = load_maf_data(maf_file_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3561f69f-7e5d-4c57-b6a8-befedcba5104",
   "metadata": {},
   "outputs": [],
   "source": [
    "def log2(value):\n",
    "    return np.log2(value + 1)\n",
    "\n",
    "def load_txt_file_into_dataframe(file_path):\n",
    "    # Read the .txt file into a pandas DataFrame\n",
    "    df = pd.read_csv(file_path, sep='\\t')  # Adjust the separator if needed\n",
    "\n",
    "    # take log2 of expression data to scale expression data\n",
    "    df = df.map(log2)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "# Call the load_txt_file_into_dataframe function\n",
    "expression_df = load_txt_file_into_dataframe(expression_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c957ea88-e994-4193-9e16-35cb3c73ba59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reformat_expression_data(df):\n",
    "    # Combine column names and index names into rows for every element\n",
    "    melted_df = pd.melt(df.reset_index(), id_vars=['index'], var_name='column', col_level=0)\n",
    "\n",
    "    # Rename columns\n",
    "    melted_df.rename(columns={'index': 'gene', 'column': 'sample', 'value': 'gene_expression'}, inplace=True)\n",
    "\n",
    "    return melted_df\n",
    "\n",
    "\n",
    "expression_df_melted = reformat_expression_data(expression_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d348536-277e-465d-b461-25a9035ff9ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# note individuals with muttated data mmay be missing in the expression data and vice versa. \n",
    "# every sample should have at least one mutation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3372f141-9ba1-4e9f-9aaf-adcf16eb43c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fraction of rows filtered is 0.2865853658536585\n"
     ]
    }
   ],
   "source": [
    "def preprocess_and_combine_mutation_expression(maf_df, expression_df):\n",
    "    ''' filter the expression data to those that have whole genome sequencing \n",
    "     i.e. appear in the mutation data frame (maf)\n",
    "     join mutation and expression data\n",
    "     '''\n",
    "    \n",
    "    exon_seq_samples = maf_df['sample'].unique()\n",
    "    filtered_expression_df = expression_df[expression_df['sample'].isin(exon_seq_samples)]\n",
    "    \n",
    "    all_rows = expression_df.shape[0]\n",
    "    filt_rows = filtered_expression_df.shape[0]\n",
    "    \n",
    "    percentage_filtered = (all_rows - filt_rows) / all_rows\n",
    "    \n",
    "    print('fraction of rows filtered is', percentage_filtered)\n",
    "\n",
    "    express_mut_genes_df = pd.merge(maf_df, filtered_expression_df, on=['gene', 'sample'], how='right')\n",
    "\n",
    "    express_mut_genes_df['mutation'].fillna(0, inplace=True)\n",
    "\n",
    "    return express_mut_genes_df\n",
    "\n",
    "\n",
    "mutation_expression_df_melted =  preprocess_and_combine_mutation_expression(maf_df, expression_df_melted)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "3a6df317-d9bc-492d-8c4f-03186375ada3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_log_fold(list_mutated, list_non_mutated, smoothing_factor=1e-8): \n",
    "    mean_mutated = mean(list_mutated)  + smoothing_factor\n",
    "    mean_non_mutated = mean(list_non_mutated) + smoothing_factor\n",
    "\n",
    "    # if np.isclose(mean_non_mutated, 0):\n",
    "    #     print(f'the denominator is zero in log2 with a value of {mean_non_mutated}')\n",
    "    #     return np.nan\n",
    "        \n",
    "    return np.log2(mean_mutated/mean_non_mutated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "377f6fb9-7597-411b-a810-af33c7f829eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_z_score(list_mutated, list_non_mutated):\n",
    "    mean_mutated = mean(list_mutated) \n",
    "    mean_non_mutated = mean(list_non_mutated) \n",
    "    all_expression_values = list_non_mutated + list_mutated\n",
    "    z_score = (mean_mutated - mean_non_mutated) / np.std(all_expression_values)\n",
    "    return z_score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "7aa8e148-4bb5-4bff-9cb3-7ae71e4bb2ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_pvalue_and_effect_size_wilcox_ranksum(list_mutated, list_non_mutated):\n",
    "     test = ranksums(list_mutated, list_non_mutated)\n",
    "     u_statistic = test.statistic \n",
    "     effect_size = u_statistic / (len(list_mutated) * len(list_non_mutated))\n",
    "     pvalue=test.pvalue\n",
    "     return pvalue, effect_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "0851fd78-7875-49c2-ace0-98a8be179d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_adjusted_pvalue(pvalues, method='fdr_bh'):\n",
    "    _, corrected_pvalues, _, _ = multitest.multipletests(pvalues, method=method)\n",
    "    return corrected_pvalues\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "231c6d1e-4a02-45bf-943c-b8ddb9c76757",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "def generate_stats_per_gene(express_mut_genes_df, target_gene, output_folder=target_gene):\n",
    "    \"\"\"\n",
    "    Given expression and mutation data calculates LogFC, pvalue, mean of expression per gene \n",
    "    for a given target_gene \n",
    "    \n",
    "    \"\"\"\n",
    "    gene_df = express_mut_genes_df[express_mut_genes_df['gene']==target_gene]\n",
    "\n",
    "    mutated_samples = gene_df[gene_df['mutation'] == 1]['sample']\n",
    "    non_mutated_samples = gene_df[gene_df['mutation'] == 0]['sample']\n",
    "    \n",
    "    mutated_individuals_expression = express_mut_genes_df[express_mut_genes_df['sample'].isin(mutated_samples)]\n",
    "    \n",
    "    non_mutated_individuals_expression = express_mut_genes_df[express_mut_genes_df['sample'].isin(non_mutated_samples)]\n",
    "\n",
    "    # gather data of mutated and non-mutated genes into lists\n",
    "    mutated_individuals_data = mutated_individuals_expression.groupby(['gene'])['gene_expression'].apply(\n",
    "    lambda x: list(x)).to_frame().reset_index().rename(columns={'gene_expression': 'gene_expression_mutated'})\n",
    "    non_mutated_individuals_data = non_mutated_individuals_expression.groupby(['gene'])['gene_expression'].apply(\n",
    "    lambda x: list(x)).to_frame().reset_index().rename(columns={'gene_expression': 'gene_expression_non_mutated'})\n",
    "\n",
    "    # combine mutated and unmutated data into one df\n",
    "    combined_data= pd.merge(mutated_individuals_data, non_mutated_individuals_data, on='gene', how='inner')\n",
    "\n",
    "    # calculate fold change and p-value and z-score\n",
    "    combined_data['logFC']= combined_data.apply(\n",
    "        lambda x: calculate_log_fold(x.gene_expression_mutated, x.gene_expression_non_mutated), axis=1)\n",
    "    # Assuming combined_data is a DataFrame\n",
    "    combined_data[['pvalue', 'effect_size']] = combined_data.apply(\n",
    "    lambda x: pd.Series(calculate_pvalue_and_effect_size_wilcox_ranksum(x.gene_expression_mutated, x.gene_expression_non_mutated)),\n",
    "    axis=1\n",
    "    )\n",
    "    combined_data['expression_mutated_mean'] = combined_data['gene_expression_mutated'].apply(mean)\n",
    "    combined_data['expression_nonmutated_mean'] = combined_data['gene_expression_non_mutated'].apply(mean)\n",
    " \n",
    "    # Clip outliers in 'logFC' column\n",
    "    # combined_data['logFC'] = combined_data['logFC'].clip(lower=-10, upper=10)\n",
    "\n",
    "    # calculate adjusted p-value\n",
    "    combined_data['adjusted_pvalue'] = calculate_adjusted_pvalue(combined_data['pvalue'].values)\n",
    "\n",
    "    # Output data to csv \n",
    "    combined_data.drop(columns=['gene_expression_mutated', 'gene_expression_non_mutated'], inplace=True)\n",
    "    \n",
    "    os.makedirs(output_folder, exist_ok=True)\n",
    "    output_filename = f'{output_folder}/{mutated_samples.count()}_{non_mutated_samples.count()}_logfc_pvalue.csv'\n",
    "    \n",
    "    print(f\"outputting data to {output_filename}\")\n",
    "    combined_data.to_csv(output_filename, index=False)\n",
    "\n",
    "    return combined_data, mutated_samples, output_filename\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "ab65d785-dbfe-4336-a4d0-68024188234c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputting data to RPL10/5_112_logfc_pvalue.csv\n"
     ]
    }
   ],
   "source": [
    "volcano_plot_df, individuals_mutated_target_gene, volcano_input_filename = generate_stats_per_gene(\n",
    "    mutation_expression_df_melted, \n",
    "    target_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "fdbd24a8-2584-4f26-9f11-48460518dd8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>expression_mutated_mean</th>\n",
       "      <th>expression_nonmutated_mean</th>\n",
       "      <th>adjusted_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>-0.357606</td>\n",
       "      <td>0.526496</td>\n",
       "      <td>-0.001131</td>\n",
       "      <td>1.796602</td>\n",
       "      <td>2.301983</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.419188</td>\n",
       "      <td>0.589864</td>\n",
       "      <td>0.000963</td>\n",
       "      <td>0.770841</td>\n",
       "      <td>0.576470</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>-0.058985</td>\n",
       "      <td>0.496169</td>\n",
       "      <td>-0.001215</td>\n",
       "      <td>7.180720</td>\n",
       "      <td>7.480391</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>0.544969</td>\n",
       "      <td>0.140050</td>\n",
       "      <td>0.002635</td>\n",
       "      <td>1.805403</td>\n",
       "      <td>1.237434</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>0.183454</td>\n",
       "      <td>0.761733</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>0.331263</td>\n",
       "      <td>0.291708</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19996</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>0.025356</td>\n",
       "      <td>0.898133</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>5.923805</td>\n",
       "      <td>5.820600</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19997</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>0.647638</td>\n",
       "      <td>0.035533</td>\n",
       "      <td>0.003754</td>\n",
       "      <td>1.981378</td>\n",
       "      <td>1.264763</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19998</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>0.145653</td>\n",
       "      <td>0.188882</td>\n",
       "      <td>0.002346</td>\n",
       "      <td>3.902844</td>\n",
       "      <td>3.528054</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19999</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>-0.133048</td>\n",
       "      <td>0.097413</td>\n",
       "      <td>-0.002960</td>\n",
       "      <td>7.000198</td>\n",
       "      <td>7.676475</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20000</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>0.010536</td>\n",
       "      <td>0.813567</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>6.142196</td>\n",
       "      <td>6.097503</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20001 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene     logFC    pvalue  effect_size  expression_mutated_mean  \\\n",
       "0         A1BG -0.357606  0.526496    -0.001131                 1.796602   \n",
       "1         A1CF  0.419188  0.589864     0.000963                 0.770841   \n",
       "2          A2M -0.058985  0.496169    -0.001215                 7.180720   \n",
       "3        A2ML1  0.544969  0.140050     0.002635                 1.805403   \n",
       "4      A3GALT2  0.183454  0.761733     0.000541                 0.331263   \n",
       "...        ...       ...       ...          ...                      ...   \n",
       "19996     ZXDC  0.025356  0.898133     0.000229                 5.923805   \n",
       "19997   ZYG11A  0.647638  0.035533     0.003754                 1.981378   \n",
       "19998   ZYG11B  0.145653  0.188882     0.002346                 3.902844   \n",
       "19999      ZYX -0.133048  0.097413    -0.002960                 7.000198   \n",
       "20000    ZZEF1  0.010536  0.813567     0.000421                 6.142196   \n",
       "\n",
       "       expression_nonmutated_mean  adjusted_pvalue  \n",
       "0                        2.301983         0.996941  \n",
       "1                        0.576470         0.996941  \n",
       "2                        7.480391         0.996941  \n",
       "3                        1.237434         0.996941  \n",
       "4                        0.291708         0.996941  \n",
       "...                           ...              ...  \n",
       "19996                    5.820600         0.996941  \n",
       "19997                    1.264763         0.996941  \n",
       "19998                    3.528054         0.996941  \n",
       "19999                    7.676475         0.996941  \n",
       "20000                    6.097503         0.996941  \n",
       "\n",
       "[20001 rows x 7 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "volcano_plot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9060c9cf-96f2-4094-8a66-70ca9b544e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot_df['effect_size'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "72c7bdd0-244a-4080-8175-3bfe47c6f1bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>expression_mutated_mean</th>\n",
       "      <th>expression_nonmutated_mean</th>\n",
       "      <th>adjusted_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>-0.357606</td>\n",
       "      <td>0.526496</td>\n",
       "      <td>-0.001131</td>\n",
       "      <td>1.796602</td>\n",
       "      <td>2.301983</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.419188</td>\n",
       "      <td>0.589864</td>\n",
       "      <td>0.000963</td>\n",
       "      <td>0.770841</td>\n",
       "      <td>0.576470</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>-0.058985</td>\n",
       "      <td>0.496169</td>\n",
       "      <td>-0.001215</td>\n",
       "      <td>7.180720</td>\n",
       "      <td>7.480391</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>0.544969</td>\n",
       "      <td>0.140050</td>\n",
       "      <td>0.002635</td>\n",
       "      <td>1.805403</td>\n",
       "      <td>1.237434</td>\n",
       "      <td>0.996941</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene     logFC    pvalue  effect_size  expression_mutated_mean  \\\n",
       "0   A1BG -0.357606  0.526496    -0.001131                 1.796602   \n",
       "1   A1CF  0.419188  0.589864     0.000963                 0.770841   \n",
       "2    A2M -0.058985  0.496169    -0.001215                 7.180720   \n",
       "3  A2ML1  0.544969  0.140050     0.002635                 1.805403   \n",
       "\n",
       "   expression_nonmutated_mean  adjusted_pvalue  \n",
       "0                    2.301983         0.996941  \n",
       "1                    0.576470         0.996941  \n",
       "2                    7.480391         0.996941  \n",
       "3                    1.237434         0.996941  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "volcano_plot_df.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "ed4500d4-0219-4d60-bf6a-a8df7ce6843d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>expression_mutated_mean</th>\n",
       "      <th>expression_nonmutated_mean</th>\n",
       "      <th>adjusted_pvalue</th>\n",
       "      <th>abs_logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14559</th>\n",
       "      <td>RNU6-602P</td>\n",
       "      <td>26.651377</td>\n",
       "      <td>0.450461</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>1.054056</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>26.651377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10679</th>\n",
       "      <td>NPB</td>\n",
       "      <td>-25.721730</td>\n",
       "      <td>0.129511</td>\n",
       "      <td>-0.002707</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.553366</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>25.721730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18936</th>\n",
       "      <td>VTRNA1-1</td>\n",
       "      <td>-25.495009</td>\n",
       "      <td>0.736195</td>\n",
       "      <td>-0.000602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.472893</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>25.495009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10836</th>\n",
       "      <td>NT5DC4</td>\n",
       "      <td>-25.263995</td>\n",
       "      <td>0.113328</td>\n",
       "      <td>-0.002828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.402921</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>25.263995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14485</th>\n",
       "      <td>RNU4-40P</td>\n",
       "      <td>25.159152</td>\n",
       "      <td>0.450461</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>0.374679</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>25.159152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2423</th>\n",
       "      <td>CCL24</td>\n",
       "      <td>-25.105041</td>\n",
       "      <td>0.121209</td>\n",
       "      <td>-0.002767</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.360886</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>25.105041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14569</th>\n",
       "      <td>RNU6-761P</td>\n",
       "      <td>24.986909</td>\n",
       "      <td>0.450461</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>0.332513</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.986909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4253</th>\n",
       "      <td>DLK1</td>\n",
       "      <td>-24.837117</td>\n",
       "      <td>0.177792</td>\n",
       "      <td>-0.002406</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.299721</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.837117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>CTAG2</td>\n",
       "      <td>-24.836412</td>\n",
       "      <td>0.438422</td>\n",
       "      <td>-0.001384</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.299575</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.836412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8447</th>\n",
       "      <td>KRTAP20-3</td>\n",
       "      <td>-24.827145</td>\n",
       "      <td>0.312166</td>\n",
       "      <td>-0.001805</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.297656</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.827145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10553</th>\n",
       "      <td>NKX2-8</td>\n",
       "      <td>-24.809185</td>\n",
       "      <td>0.050702</td>\n",
       "      <td>-0.003489</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.293974</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.809185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8468</th>\n",
       "      <td>KRTAP4-3</td>\n",
       "      <td>-24.807331</td>\n",
       "      <td>0.177792</td>\n",
       "      <td>-0.002406</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.293596</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.807331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14700</th>\n",
       "      <td>RPL3L</td>\n",
       "      <td>-24.744327</td>\n",
       "      <td>0.157080</td>\n",
       "      <td>-0.002527</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.281051</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.744327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9689</th>\n",
       "      <td>MOG</td>\n",
       "      <td>-24.566484</td>\n",
       "      <td>0.092089</td>\n",
       "      <td>-0.003008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.248455</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.566484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11569</th>\n",
       "      <td>PADI4</td>\n",
       "      <td>-24.521875</td>\n",
       "      <td>0.113328</td>\n",
       "      <td>-0.002828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.240891</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.521875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7377</th>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>-24.479748</td>\n",
       "      <td>0.946280</td>\n",
       "      <td>-0.000120</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.233958</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.479748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7513</th>\n",
       "      <td>IGLV3-12</td>\n",
       "      <td>-24.474196</td>\n",
       "      <td>0.238344</td>\n",
       "      <td>-0.002106</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.233060</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.474196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18028</th>\n",
       "      <td>TRBV5-4</td>\n",
       "      <td>-24.448005</td>\n",
       "      <td>0.157080</td>\n",
       "      <td>-0.002527</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.228867</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.448005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2105</th>\n",
       "      <td>CALML5</td>\n",
       "      <td>-24.404493</td>\n",
       "      <td>0.418774</td>\n",
       "      <td>-0.001444</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222067</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.404493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11956</th>\n",
       "      <td>PF4V1</td>\n",
       "      <td>-24.402813</td>\n",
       "      <td>0.212576</td>\n",
       "      <td>-0.002226</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.221809</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.402813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8298</th>\n",
       "      <td>KLK6</td>\n",
       "      <td>-24.401086</td>\n",
       "      <td>0.113328</td>\n",
       "      <td>-0.002828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.221543</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.401086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14578</th>\n",
       "      <td>RNU6ATAC10P</td>\n",
       "      <td>-24.390973</td>\n",
       "      <td>0.710946</td>\n",
       "      <td>-0.000662</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.219996</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.390973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8445</th>\n",
       "      <td>KRTAP20-1</td>\n",
       "      <td>-24.360393</td>\n",
       "      <td>0.566832</td>\n",
       "      <td>-0.001023</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.215382</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.360393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14084</th>\n",
       "      <td>RN7SL454P</td>\n",
       "      <td>24.358865</td>\n",
       "      <td>0.450461</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>0.215154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.358865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18032</th>\n",
       "      <td>TRBV6-1</td>\n",
       "      <td>-24.325627</td>\n",
       "      <td>0.399654</td>\n",
       "      <td>-0.001504</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.210254</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.325627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11539</th>\n",
       "      <td>P3R3URF</td>\n",
       "      <td>-24.262919</td>\n",
       "      <td>0.522106</td>\n",
       "      <td>-0.001143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.201310</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.262919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1884</th>\n",
       "      <td>C20orf141</td>\n",
       "      <td>-24.252362</td>\n",
       "      <td>0.345523</td>\n",
       "      <td>-0.001684</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.199843</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.252362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3139</th>\n",
       "      <td>CLEC2L</td>\n",
       "      <td>-24.235763</td>\n",
       "      <td>0.312166</td>\n",
       "      <td>-0.001805</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.197557</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.235763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7514</th>\n",
       "      <td>IGLV3-16</td>\n",
       "      <td>-24.193196</td>\n",
       "      <td>0.328570</td>\n",
       "      <td>-0.001745</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.191813</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.193196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19114</th>\n",
       "      <td>XAGE1B</td>\n",
       "      <td>-24.178299</td>\n",
       "      <td>0.661412</td>\n",
       "      <td>-0.000782</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.189842</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.178299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8297</th>\n",
       "      <td>KLK5</td>\n",
       "      <td>-24.167875</td>\n",
       "      <td>0.266243</td>\n",
       "      <td>-0.001985</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.188476</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.167875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19113</th>\n",
       "      <td>XAGE1A</td>\n",
       "      <td>-24.166411</td>\n",
       "      <td>0.686011</td>\n",
       "      <td>-0.000722</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.188284</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.166411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>ASB10</td>\n",
       "      <td>-24.155666</td>\n",
       "      <td>0.238344</td>\n",
       "      <td>-0.002106</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.186887</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.155666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10138</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>-24.155059</td>\n",
       "      <td>0.566832</td>\n",
       "      <td>-0.001023</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.186809</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.155059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891</th>\n",
       "      <td>AQP12B</td>\n",
       "      <td>-24.148844</td>\n",
       "      <td>0.438422</td>\n",
       "      <td>-0.001384</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.186006</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.148844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15254</th>\n",
       "      <td>SERPINB4</td>\n",
       "      <td>-24.120548</td>\n",
       "      <td>0.252025</td>\n",
       "      <td>-0.002045</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.182393</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.120548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4568</th>\n",
       "      <td>DYTN</td>\n",
       "      <td>-24.116844</td>\n",
       "      <td>0.252025</td>\n",
       "      <td>-0.002045</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.181925</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.116844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7384</th>\n",
       "      <td>IGHV1-45</td>\n",
       "      <td>-24.079144</td>\n",
       "      <td>0.399654</td>\n",
       "      <td>-0.001504</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.177233</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.079144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14565</th>\n",
       "      <td>RNU6-707P</td>\n",
       "      <td>-24.078271</td>\n",
       "      <td>0.761733</td>\n",
       "      <td>-0.000541</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.177126</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.078271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3601</th>\n",
       "      <td>CSAG3</td>\n",
       "      <td>-24.007893</td>\n",
       "      <td>0.188882</td>\n",
       "      <td>-0.002346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.168693</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.007893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5141</th>\n",
       "      <td>FABP1</td>\n",
       "      <td>-24.003578</td>\n",
       "      <td>0.418774</td>\n",
       "      <td>-0.001444</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.168189</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>24.003578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5824</th>\n",
       "      <td>FXYD4</td>\n",
       "      <td>-23.988692</td>\n",
       "      <td>0.399654</td>\n",
       "      <td>-0.001504</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166462</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.988692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18045</th>\n",
       "      <td>TRBV7-9</td>\n",
       "      <td>-23.942375</td>\n",
       "      <td>0.345523</td>\n",
       "      <td>-0.001684</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.161203</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.942375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14928</th>\n",
       "      <td>SAA2-SAA4</td>\n",
       "      <td>-23.942258</td>\n",
       "      <td>0.121209</td>\n",
       "      <td>-0.002767</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.161190</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.942258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2572</th>\n",
       "      <td>CD70</td>\n",
       "      <td>-23.937414</td>\n",
       "      <td>0.281004</td>\n",
       "      <td>-0.001925</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.160650</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.937414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14527</th>\n",
       "      <td>RNU6-19P</td>\n",
       "      <td>-23.932877</td>\n",
       "      <td>0.892803</td>\n",
       "      <td>-0.000241</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.160145</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.932877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17977</th>\n",
       "      <td>TRAV25</td>\n",
       "      <td>-23.925031</td>\n",
       "      <td>0.438422</td>\n",
       "      <td>-0.001384</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.159277</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.925031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8440</th>\n",
       "      <td>KRTAP19-8</td>\n",
       "      <td>-23.921568</td>\n",
       "      <td>0.661412</td>\n",
       "      <td>-0.000782</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.158895</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.921568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13327</th>\n",
       "      <td>RAET1L</td>\n",
       "      <td>-23.921211</td>\n",
       "      <td>0.129511</td>\n",
       "      <td>-0.002707</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.158855</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.921211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3576</th>\n",
       "      <td>CRYAA</td>\n",
       "      <td>-23.912769</td>\n",
       "      <td>0.188882</td>\n",
       "      <td>-0.002346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.157929</td>\n",
       "      <td>0.996941</td>\n",
       "      <td>23.912769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene      logFC    pvalue  effect_size  expression_mutated_mean  \\\n",
       "14559    RNU6-602P  26.651377  0.450461     0.001348                 1.054056   \n",
       "10679          NPB -25.721730  0.129511    -0.002707                 0.000000   \n",
       "18936     VTRNA1-1 -25.495009  0.736195    -0.000602                 0.000000   \n",
       "10836       NT5DC4 -25.263995  0.113328    -0.002828                 0.000000   \n",
       "14485     RNU4-40P  25.159152  0.450461     0.001348                 0.374679   \n",
       "2423         CCL24 -25.105041  0.121209    -0.002767                 0.000000   \n",
       "14569    RNU6-761P  24.986909  0.450461     0.001348                 0.332513   \n",
       "4253          DLK1 -24.837117  0.177792    -0.002406                 0.000000   \n",
       "3677         CTAG2 -24.836412  0.438422    -0.001384                 0.000000   \n",
       "8447     KRTAP20-3 -24.827145  0.312166    -0.001805                 0.000000   \n",
       "10553       NKX2-8 -24.809185  0.050702    -0.003489                 0.000000   \n",
       "8468      KRTAP4-3 -24.807331  0.177792    -0.002406                 0.000000   \n",
       "14700        RPL3L -24.744327  0.157080    -0.002527                 0.000000   \n",
       "9689           MOG -24.566484  0.092089    -0.003008                 0.000000   \n",
       "11569        PADI4 -24.521875  0.113328    -0.002828                 0.000000   \n",
       "7377         IGHJ5 -24.479748  0.946280    -0.000120                 0.000000   \n",
       "7513      IGLV3-12 -24.474196  0.238344    -0.002106                 0.000000   \n",
       "18028      TRBV5-4 -24.448005  0.157080    -0.002527                 0.000000   \n",
       "2105        CALML5 -24.404493  0.418774    -0.001444                 0.000000   \n",
       "11956        PF4V1 -24.402813  0.212576    -0.002226                 0.000000   \n",
       "8298          KLK6 -24.401086  0.113328    -0.002828                 0.000000   \n",
       "14578  RNU6ATAC10P -24.390973  0.710946    -0.000662                 0.000000   \n",
       "8445     KRTAP20-1 -24.360393  0.566832    -0.001023                 0.000000   \n",
       "14084    RN7SL454P  24.358865  0.450461     0.001348                 0.215154   \n",
       "18032      TRBV6-1 -24.325627  0.399654    -0.001504                 0.000000   \n",
       "11539      P3R3URF -24.262919  0.522106    -0.001143                 0.000000   \n",
       "1884     C20orf141 -24.252362  0.345523    -0.001684                 0.000000   \n",
       "3139        CLEC2L -24.235763  0.312166    -0.001805                 0.000000   \n",
       "7514      IGLV3-16 -24.193196  0.328570    -0.001745                 0.000000   \n",
       "19114       XAGE1B -24.178299  0.661412    -0.000782                 0.000000   \n",
       "8297          KLK5 -24.167875  0.266243    -0.001985                 0.000000   \n",
       "19113       XAGE1A -24.166411  0.686011    -0.000722                 0.000000   \n",
       "1105         ASB10 -24.155666  0.238344    -0.002106                 0.000000   \n",
       "10138        MYOD1 -24.155059  0.566832    -0.001023                 0.000000   \n",
       "891         AQP12B -24.148844  0.438422    -0.001384                 0.000000   \n",
       "15254     SERPINB4 -24.120548  0.252025    -0.002045                 0.000000   \n",
       "4568          DYTN -24.116844  0.252025    -0.002045                 0.000000   \n",
       "7384      IGHV1-45 -24.079144  0.399654    -0.001504                 0.000000   \n",
       "14565    RNU6-707P -24.078271  0.761733    -0.000541                 0.000000   \n",
       "3601         CSAG3 -24.007893  0.188882    -0.002346                 0.000000   \n",
       "5141         FABP1 -24.003578  0.418774    -0.001444                 0.000000   \n",
       "5824         FXYD4 -23.988692  0.399654    -0.001504                 0.000000   \n",
       "18045      TRBV7-9 -23.942375  0.345523    -0.001684                 0.000000   \n",
       "14928    SAA2-SAA4 -23.942258  0.121209    -0.002767                 0.000000   \n",
       "2572          CD70 -23.937414  0.281004    -0.001925                 0.000000   \n",
       "14527     RNU6-19P -23.932877  0.892803    -0.000241                 0.000000   \n",
       "17977       TRAV25 -23.925031  0.438422    -0.001384                 0.000000   \n",
       "8440     KRTAP19-8 -23.921568  0.661412    -0.000782                 0.000000   \n",
       "13327       RAET1L -23.921211  0.129511    -0.002707                 0.000000   \n",
       "3576         CRYAA -23.912769  0.188882    -0.002346                 0.000000   \n",
       "\n",
       "       expression_nonmutated_mean  adjusted_pvalue  abs_logFC  \n",
       "14559                    0.000000         0.996941  26.651377  \n",
       "10679                    0.553366         0.996941  25.721730  \n",
       "18936                    0.472893         0.996941  25.495009  \n",
       "10836                    0.402921         0.996941  25.263995  \n",
       "14485                    0.000000         0.996941  25.159152  \n",
       "2423                     0.360886         0.996941  25.105041  \n",
       "14569                    0.000000         0.996941  24.986909  \n",
       "4253                     0.299721         0.996941  24.837117  \n",
       "3677                     0.299575         0.996941  24.836412  \n",
       "8447                     0.297656         0.996941  24.827145  \n",
       "10553                    0.293974         0.996941  24.809185  \n",
       "8468                     0.293596         0.996941  24.807331  \n",
       "14700                    0.281051         0.996941  24.744327  \n",
       "9689                     0.248455         0.996941  24.566484  \n",
       "11569                    0.240891         0.996941  24.521875  \n",
       "7377                     0.233958         0.996941  24.479748  \n",
       "7513                     0.233060         0.996941  24.474196  \n",
       "18028                    0.228867         0.996941  24.448005  \n",
       "2105                     0.222067         0.996941  24.404493  \n",
       "11956                    0.221809         0.996941  24.402813  \n",
       "8298                     0.221543         0.996941  24.401086  \n",
       "14578                    0.219996         0.996941  24.390973  \n",
       "8445                     0.215382         0.996941  24.360393  \n",
       "14084                    0.000000         0.996941  24.358865  \n",
       "18032                    0.210254         0.996941  24.325627  \n",
       "11539                    0.201310         0.996941  24.262919  \n",
       "1884                     0.199843         0.996941  24.252362  \n",
       "3139                     0.197557         0.996941  24.235763  \n",
       "7514                     0.191813         0.996941  24.193196  \n",
       "19114                    0.189842         0.996941  24.178299  \n",
       "8297                     0.188476         0.996941  24.167875  \n",
       "19113                    0.188284         0.996941  24.166411  \n",
       "1105                     0.186887         0.996941  24.155666  \n",
       "10138                    0.186809         0.996941  24.155059  \n",
       "891                      0.186006         0.996941  24.148844  \n",
       "15254                    0.182393         0.996941  24.120548  \n",
       "4568                     0.181925         0.996941  24.116844  \n",
       "7384                     0.177233         0.996941  24.079144  \n",
       "14565                    0.177126         0.996941  24.078271  \n",
       "3601                     0.168693         0.996941  24.007893  \n",
       "5141                     0.168189         0.996941  24.003578  \n",
       "5824                     0.166462         0.996941  23.988692  \n",
       "18045                    0.161203         0.996941  23.942375  \n",
       "14928                    0.161190         0.996941  23.942258  \n",
       "2572                     0.160650         0.996941  23.937414  \n",
       "14527                    0.160145         0.996941  23.932877  \n",
       "17977                    0.159277         0.996941  23.925031  \n",
       "8440                     0.158895         0.996941  23.921568  \n",
       "13327                    0.158855         0.996941  23.921211  \n",
       "3576                     0.157929         0.996941  23.912769  "
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "volcano_plot_df['abs_logFC'] = volcano_plot_df['logFC'].abs()\n",
    "volcano_plot_df.nlargest(50, 'abs_logFC')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e484ffc8-6249-448f-b354-01574bb19717",
   "metadata": {},
   "source": [
    "## generate heatmap of top 100 differentially expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fe591289-eba9-427a-8563-ada51b200b91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_expression_heatmap(expression_df, volcano_plot_df, n=100, exclude_value=10):\n",
    "    # Get the indices of the top n rows based on absolute values of 'logFC'\n",
    "    # get the genes most differentially expressed (high log FC values)\n",
    "    # if we exclude -exclude_value, exclude_value values\n",
    "    top_n_indices = volcano_plot_df[\n",
    "        (volcano_plot_df['logFC'].abs() != exclude_value)\n",
    "    ]['logFC'].abs().nlargest(n).index\n",
    "    \n",
    "    top_n_rows = volcano_plot_df.loc[top_n_indices].set_index('gene')\n",
    "\n",
    "    # go back to the expression df and make a heatmap of the top n genes\n",
    "    expression_df_heatmap = expression_df.loc[top_n_rows.index]\n",
    "    \n",
    "    return expression_df_heatmap\n",
    "\n",
    "heatmap_data = generate_expression_heatmap(expression_df, volcano_plot_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1bdc8bc0-fb04-4196-b9e5-45c46f3fb94b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RESPOND_10100218</th>\n",
       "      <th>RESPOND_10100291</th>\n",
       "      <th>RESPOND_10100412</th>\n",
       "      <th>RESPOND_10100478</th>\n",
       "      <th>RESPOND_10100596</th>\n",
       "      <th>RESPOND_10100615</th>\n",
       "      <th>RESPOND_10100801</th>\n",
       "      <th>RESPOND_10100884</th>\n",
       "      <th>RESPOND_10100899</th>\n",
       "      <th>RESPOND_10100952</th>\n",
       "      <th>...</th>\n",
       "      <th>RESPOND_80100242</th>\n",
       "      <th>RESPOND_80100313</th>\n",
       "      <th>RESPOND_80100345</th>\n",
       "      <th>RESPOND_80100411</th>\n",
       "      <th>RESPOND_80100526</th>\n",
       "      <th>RESPOND_80100556</th>\n",
       "      <th>RESPOND_80100590</th>\n",
       "      <th>RESPOND_81100031</th>\n",
       "      <th>RESPOND_40100842</th>\n",
       "      <th>RESPOND_80100259</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TMEM238L</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RN7SL321P</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNVU1-19</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.915644</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.471448</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 164 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           RESPOND_10100218  RESPOND_10100291  RESPOND_10100412  \\\n",
       "gene                                                              \n",
       "TMEM238L                0.0               0.0               0.0   \n",
       "RN7SL321P               0.0               0.0               0.0   \n",
       "RNVU1-19                0.0               0.0               0.0   \n",
       "\n",
       "           RESPOND_10100478  RESPOND_10100596  RESPOND_10100615  \\\n",
       "gene                                                              \n",
       "TMEM238L                0.0          0.000012               0.0   \n",
       "RN7SL321P               0.0          0.000000               0.0   \n",
       "RNVU1-19                0.0          0.000000               0.0   \n",
       "\n",
       "           RESPOND_10100801  RESPOND_10100884  RESPOND_10100899  \\\n",
       "gene                                                              \n",
       "TMEM238L           0.000003               0.0               0.0   \n",
       "RN7SL321P          0.000000               0.0               0.0   \n",
       "RNVU1-19           0.000000               0.0               0.0   \n",
       "\n",
       "           RESPOND_10100952  ...  RESPOND_80100242  RESPOND_80100313  \\\n",
       "gene                         ...                                       \n",
       "TMEM238L                0.0  ...          0.000000               0.0   \n",
       "RN7SL321P               0.0  ...          0.000000               0.0   \n",
       "RNVU1-19                0.0  ...          0.915644               0.0   \n",
       "\n",
       "           RESPOND_80100345  RESPOND_80100411  RESPOND_80100526  \\\n",
       "gene                                                              \n",
       "TMEM238L                0.0               0.0               0.0   \n",
       "RN7SL321P               0.0               0.0               0.0   \n",
       "RNVU1-19                0.0               0.0               0.0   \n",
       "\n",
       "           RESPOND_80100556  RESPOND_80100590  RESPOND_81100031  \\\n",
       "gene                                                              \n",
       "TMEM238L                0.0               0.0               0.0   \n",
       "RN7SL321P               0.0               0.0               0.0   \n",
       "RNVU1-19                0.0               0.0               0.0   \n",
       "\n",
       "           RESPOND_40100842  RESPOND_80100259  \n",
       "gene                                           \n",
       "TMEM238L           0.000025               0.0  \n",
       "RN7SL321P          0.000000               0.0  \n",
       "RNVU1-19           0.471448               0.0  \n",
       "\n",
       "[3 rows x 164 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "heatmap_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a25a4468-1792-447c-94e8-1c384c96c505",
   "metadata": {},
   "outputs": [],
   "source": [
    "# can check mutation status of genes by checking individuals_mutated_target_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e51e4364-9eec-4ef8-ba66-16410c606ac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPL10/sample_mutation_status.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_mutated_status(expression_df_heatmap, individuals_mutated_target_gene, output_folder=target_gene):\n",
    "    individuals_mutated_target_gene = list(individuals_mutated_target_gene)\n",
    "    \n",
    "    mutated_status = [1 if item in individuals_mutated_target_gene else 0 for item in expression_df_heatmap.columns]\n",
    "\n",
    "    sample_categories_df = pd.DataFrame({\n",
    "        'Sample': expression_df_heatmap.columns,\n",
    "        'Mutation Status': mutated_status\n",
    "    })\n",
    "\n",
    "    output_filename = f'{output_folder}/sample_mutation_status.csv'\n",
    "    print(output_filename)\n",
    "    sample_categories_df.to_csv(output_filename,  index=False)\n",
    "    return sample_categories_df\n",
    "\n",
    "mutated_status_df = get_mutated_status(heatmap_data, individuals_mutated_target_gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "74d85b53-9bb0-478f-a08d-d7d0f8186c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.cluster.hierarchy import fcluster, linkage\n",
    "\n",
    "def hierarchical_clustering(expression_df_heatmap, row_threshold=7, col_threshold=7, output_folder=target_gene):\n",
    "    # Cluster the rows and columns using hierarchical clustering\n",
    "    row_linkage = linkage(expression_df_heatmap, method='ward')\n",
    "    col_linkage = linkage(expression_df_heatmap.T, method='ward')\n",
    "\n",
    "    # Assign cluster labels using cluster\n",
    "    row_clusters = fcluster(row_linkage, t=row_threshold, criterion='maxclust')\n",
    "    col_clusters = fcluster(col_linkage, t=col_threshold, criterion='maxclust')\n",
    "\n",
    "    # Create dictionaries to store cluster information\n",
    "    row_cluster_info = {f\"Cluster {cluster}\": expression_df_heatmap.index[row_clusters == cluster] for cluster in np.unique(row_clusters)}\n",
    "    col_cluster_info = {f\"Cluster {cluster}\": expression_df_heatmap.columns[col_clusters == cluster] for cluster in np.unique(col_clusters)}\n",
    "\n",
    "    row_cluster_file = os.path.join(output_folder, 'row_clusters.txt')\n",
    "    with open(row_cluster_file, 'w') as file:\n",
    "        for cluster, items in row_cluster_info.items():\n",
    "            file.write(f\"{cluster}:\\n\")\n",
    "            file.write(f\"{', '.join(items)}\\n\\n\")\n",
    "\n",
    "    # Save column cluster information to a text file\n",
    "    col_cluster_file = os.path.join(output_folder, 'col_clusters.txt')\n",
    "    with open(col_cluster_file, 'w') as file:\n",
    "        for cluster, items in col_cluster_info.items():\n",
    "            file.write(f\"{cluster}:\\n\")\n",
    "            file.write(f\"{', '.join(items)}\\n\\n\")\n",
    "    \n",
    "    return row_linkage, col_linkage, row_cluster_info, col_cluster_info\n",
    "\n",
    "# Example usage:\n",
    "row_linkage, col_linkage, row_clusters, col_clusters = hierarchical_clustering(heatmap_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "59d7b1f7-1930-45d1-bab9-b2cab98971fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def create_clustered_heatmap_and_save(expression_df_heatmap, \n",
    "                                      row_linkage, \n",
    "                                      col_linkage, \n",
    "                                      sample_mutation_df, \n",
    "                                      output_folder=target_gene):\n",
    "    # Group labels and colors for color bar\n",
    "    type_map = {1: 'red', 0: 'yellow'}\n",
    "\n",
    "    # Create a clustered heatmap\n",
    "    clustered_df = sns.clustermap(\n",
    "        expression_df_heatmap,\n",
    "        row_linkage=row_linkage,\n",
    "        col_linkage=col_linkage,\n",
    "        cmap='viridis',\n",
    "        annot=True,\n",
    "        fmt=\".1f\",\n",
    "        linewidths=.5,\n",
    "        col_colors=sample_mutation_df.set_index('Sample')['Mutation Status'].map(type_map),\n",
    "        standard_scale=0  # standardize by the rows\n",
    "    )\n",
    "    output_filename = f'{output_folder}/heatmap.png'\n",
    "    \n",
    "    # Save the plot to a file\n",
    "    clustered_df.savefig(output_filename)\n",
    "\n",
    "    # Close the plot to prevent displaying it in the notebook (optional)\n",
    "    plt.close()\n",
    "\n",
    "    return clustered_df\n",
    "\n",
    "clustered_heatmap = create_clustered_heatmap_and_save(heatmap_data, \n",
    "                                                      row_linkage, \n",
    "                                                      col_linkage,\n",
    "                                                      mutated_status_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "56c644ef-b34e-4343-b27e-e281a2be9aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram\n",
    "\n",
    "## add more args to the function \n",
    "\n",
    "def plot_and_save_dendrograms(row_linkage, col_linkage, expression_df_heatmap, output_folder=target_gene):\n",
    "    # Plot the row dendrogram\n",
    "    plt.figure(figsize=(12, 6))\n",
    "    gene_plot = dendrogram(row_linkage, \n",
    "                           labels=expression_df_heatmap.index, \n",
    "                           truncate_mode='lastp', \n",
    "                           p=int(0.75 * len(row_linkage)), show_leaf_counts=True)\n",
    "    plt.title('Row Dendrogram')\n",
    "    \n",
    "    # Save the row dendrogram plot to a file\n",
    "    row_dendrogram_filename = f'{output_folder}/row_dendrogram.png'\n",
    "    plt.savefig(row_dendrogram_filename)\n",
    "    plt.close()\n",
    "\n",
    "    # Plot the column dendrogram\n",
    "    plt.figure(figsize=(12, 6))\n",
    "    sample_plot = dendrogram(col_linkage, labels=expression_df_heatmap.columns, orientation='top', \n",
    "                             truncate_mode='lastp', p=int(0.75 * len(row_linkage)), show_leaf_counts=True)\n",
    "    plt.title('Column Dendrogram')\n",
    "    \n",
    "    # Save the column dendrogram plot to a file\n",
    "    col_dendrogram_filename = f'{output_folder}/col_dendrogram.png'\n",
    "    plt.savefig(col_dendrogram_filename)\n",
    "    plt.close()\n",
    "\n",
    "    return gene_plot, sample_plot\n",
    "\n",
    "row_dendrogram, col_dendrogram = plot_and_save_dendrograms(row_linkage, \n",
    "                                                           col_linkage, \n",
    "                                                           heatmap_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0e7b9c4a-c204-40a4-af6f-a80cbb4a0594",
   "metadata": {},
   "outputs": [],
   "source": [
    "def histogram_of_column_and_save(df, column, output_folder=target_gene):\n",
    "    # Plot the distribution of p-values\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.hist(df[column], bins=30, color='blue', edgecolor='black')\n",
    "    plt.title(f'Distribution of {column}')\n",
    "    plt.xlabel('P-values')\n",
    "    plt.ylabel('Frequency')\n",
    "\n",
    "    output_filename = f'{output_folder}/histogram_{column}.png'\n",
    "    # Save the histogram plot to a file\n",
    "    plt.savefig(output_filename)\n",
    "    plt.close()\n",
    "\n",
    "histogram_of_column_and_save(volcano_plot_df, 'adjusted_pvalue')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0d760684-969d-4d52-953b-592e349083d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clip logFC for the plot. not above bc need the full value for the histogram \n",
    "def volcano_plot(input_file_path, yaxis, xaxis, output_folder=target_gene):\n",
    "    df = pd.read_csv(input_file_path)\n",
    "    plt.scatter(x=df[xaxis],y=df[yaxis].apply(lambda x:-np.log10(x)),s=1)\n",
    "\n",
    "    plt.xlabel(xaxis)\n",
    "    plt.ylabel(f'-log10({yaxis})')\n",
    "    plt.title('Volcano Plot')\n",
    "    output_filename = f'{output_folder}/volcano_plot.png'\n",
    "    plt.savefig(output_filename)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a35e2751-dd01-4e90-a758-433602a430e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_expression_boxplots(expression_df, volcano_plot_df, n=10, exclude_value=10):\n",
    "    # Get the indices of the top n rows based on absolute values of 'logFC'\n",
    "    # get the genes most differentially expressed (high log FC values)\n",
    "    # if we exclude -exclude_value, exclude_value values\n",
    "    top_n_indices = volcano_plot_df[\n",
    "        (volcano_plot_df['logFC'].abs() != exclude_value)\n",
    "    ]['logFC'].abs().nlargest(n).index\n",
    "    \n",
    "    top_n_rows = volcano_plot_df.loc[top_n_indices].set_index('gene')\n",
    "\n",
    "    # go back to the expression df and make a heatmap of the top n genes\n",
    "    expression_df_heatmap = expression_df.loc[top_n_rows.index]\n",
    "    \n",
    "    return expression_df_heatmap\n",
    "\n",
    "heatmap_data = generate_expression_heatmap(expression_df, volcano_plot_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f006b5df-5b17-47a0-8f43-03a9b9e52cc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(input_file_path=volcano_input_filename, yaxis='pvalue', xaxis='logFC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb82f15e-2d3f-4c9d-b292-99ccadc4b8ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b4dfda-1b74-4630-ab43-d4e4fa9a7490",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
