{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a68179f1-6218-4af0-b352-79dccdb72e82",
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_file_path = 'data/RESPOND_247_coding_final.maf'\n",
    "expression_file_path = 'data/Expression_remove_BE.txt'\n",
    "target_gene='SPOP'\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statistics import mean \n",
    "from scipy.stats import ranksums\n",
    "import matplotlib.pylab as plt\n",
    "from statsmodels.stats import multitest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "12bdca97-76e8-4627-8c49-51098cc5fd8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_maf_data(file_path):\n",
    "    # Define columns of interest\n",
    "    columns = [\"Hugo_Symbol\", \"Tumor_Sample_Barcode\"]\n",
    "    df = pd.read_csv(file_path, sep='\\t', comment=\"#\", usecols=columns)\n",
    "    df.rename(columns={'Hugo_Symbol': 'gene', 'Tumor_Sample_Barcode': 'sample'}, inplace=True)\n",
    "    df['mutation'] = 1\n",
    "    return df\n",
    "    \n",
    "maf_df = load_maf_data(maf_file_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "3561f69f-7e5d-4c57-b6a8-befedcba5104",
   "metadata": {},
   "outputs": [],
   "source": [
    "def log2(value):\n",
    "    return np.log2(value + 1)\n",
    "\n",
    "def load_txt_file_into_dataframe(file_path):\n",
    "    # Read the .txt file into a pandas DataFrame\n",
    "    df = pd.read_csv(file_path, sep='\\t')  # Adjust the separator if needed\n",
    "\n",
    "    # take log2 of expression data to scale expression data\n",
    "    df = df.map(log2)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "# Call the load_txt_file_into_dataframe function\n",
    "expression_df = load_txt_file_into_dataframe(expression_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "c957ea88-e994-4193-9e16-35cb3c73ba59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reformat_expression_data(df):\n",
    "    # Combine column names and index names into rows for every element\n",
    "    melted_df = pd.melt(df.reset_index(), id_vars=['index'], var_name='column', col_level=0)\n",
    "\n",
    "    # Rename columns\n",
    "    melted_df.rename(columns={'index': 'gene', 'column': 'sample', 'value': 'gene_expression'}, inplace=True)\n",
    "\n",
    "    return melted_df\n",
    "\n",
    "\n",
    "expression_df_melted = reformat_expression_data(expression_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "6d348536-277e-465d-b461-25a9035ff9ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# note individuals with muttated data may be missing in the expression data and vice versa. \n",
    "# every sample should have at least one mutation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "3372f141-9ba1-4e9f-9aaf-adcf16eb43c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fraction of rows filtered is 0.2865853658536585\n"
     ]
    }
   ],
   "source": [
    "def preprocess_and_combine_mutation_expression(maf_df, expression_df):\n",
    "    ''' filter the expression data to those that have whole genome sequencing \n",
    "     i.e. appear in the mutation data frame (maf)\n",
    "     join mutation and expression data\n",
    "     '''\n",
    "    \n",
    "    exon_seq_samples = maf_df['sample'].unique()\n",
    "    filtered_expression_df = expression_df[expression_df['sample'].isin(exon_seq_samples)]\n",
    "    \n",
    "    all_rows = expression_df.shape[0]\n",
    "    filt_rows = filtered_expression_df.shape[0]\n",
    "    \n",
    "    percentage_filtered = (all_rows - filt_rows) / all_rows\n",
    "    \n",
    "    print('fraction of rows filtered is', percentage_filtered)\n",
    "\n",
    "    express_mut_genes_df = pd.merge(maf_df, filtered_expression_df, on=['gene', 'sample'], how='right')\n",
    "\n",
    "    express_mut_genes_df['mutation'].fillna(0, inplace=True)\n",
    "\n",
    "    return express_mut_genes_df\n",
    "\n",
    "\n",
    "mutation_expression_df_melted =  preprocess_and_combine_mutation_expression(maf_df, expression_df_melted)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "3a6df317-d9bc-492d-8c4f-03186375ada3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_log_fold(list_mutated, list_non_mutated, smoothing_factor=1e-8): \n",
    "    mean_mutated = mean(list_mutated)  + smoothing_factor\n",
    "    mean_non_mutated = mean(list_non_mutated) + smoothing_factor\n",
    "        \n",
    "    return np.log2(mean_mutated/mean_non_mutated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "377f6fb9-7597-411b-a810-af33c7f829eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_z_score(list_mutated, list_non_mutated):\n",
    "    mean_mutated = mean(list_mutated) \n",
    "    mean_non_mutated = mean(list_non_mutated) \n",
    "    all_expression_values = list_non_mutated + list_mutated\n",
    "    z_score = (mean_mutated - mean_non_mutated) / np.std(all_expression_values)\n",
    "    return z_score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "36736785-ca3e-4a5e-8b34-d550cef0d3ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this isn't used for gene expression \n",
    "# from scipy.stats import ttest_ind\n",
    "# def calculate_pvalue_and_effect_size_ttest(list_mutated, list_non_mutated):\n",
    "#     t_stat, p_value = ttest_ind(list_mutated, list_non_mutated, equal_var=False)\n",
    "#     return p_value, t_stat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "7aa8e148-4bb5-4bff-9cb3-7ae71e4bb2ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_pvalue_and_effect_size_wilcox_ranksum(list_mutated, list_non_mutated):\n",
    "     test = ranksums(list_mutated, list_non_mutated)\n",
    "     u_statistic = test.statistic \n",
    "     effect_size = u_statistic / (len(list_mutated) * len(list_non_mutated))\n",
    "     pvalue=test.pvalue\n",
    "     return pvalue, effect_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "0851fd78-7875-49c2-ace0-98a8be179d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_adjusted_pvalue(pvalues, method='fdr_bh'):\n",
    "    _, corrected_pvalues, _, _ = multitest.multipletests(pvalues, method=method)\n",
    "    return corrected_pvalues\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "78f10ce4-a336-4741-944c-7642a31e304f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SPOP'"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "231c6d1e-4a02-45bf-943c-b8ddb9c76757",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "def generate_stats_per_gene(express_mut_genes_df, target_gene, output_folder=target_gene):\n",
    "    \"\"\"\n",
    "    Given expression and mutation data calculates LogFC, pvalue, mean of expression per gene \n",
    "    for a given target_gene \n",
    "    \n",
    "    \"\"\"\n",
    "    gene_df = express_mut_genes_df[express_mut_genes_df['gene']==target_gene]\n",
    "\n",
    "    if len(gene_df) == 0:\n",
    "        raise ValueError(\"This gene is not valid! No mutations with this gene exist\") \n",
    "\n",
    "    mutated_samples = gene_df[gene_df['mutation'] == 1]['sample']\n",
    "    non_mutated_samples = gene_df[gene_df['mutation'] == 0]['sample']\n",
    "    \n",
    "    mutated_individuals_expression = express_mut_genes_df[express_mut_genes_df['sample'].isin(mutated_samples)]\n",
    "    \n",
    "    non_mutated_individuals_expression = express_mut_genes_df[express_mut_genes_df['sample'].isin(non_mutated_samples)]\n",
    "\n",
    "    # gather data of mutated and non-mutated genes into lists\n",
    "    mutated_individuals_data = mutated_individuals_expression.groupby(['gene'])['gene_expression'].apply(\n",
    "    lambda x: list(x)).to_frame().reset_index().rename(columns={'gene_expression': 'gene_expression_mutated'})\n",
    "    non_mutated_individuals_data = non_mutated_individuals_expression.groupby(['gene'])['gene_expression'].apply(\n",
    "    lambda x: list(x)).to_frame().reset_index().rename(columns={'gene_expression': 'gene_expression_non_mutated'})\n",
    "\n",
    "    # combine mutated and unmutated data into one df\n",
    "    combined_data= pd.merge(mutated_individuals_data, non_mutated_individuals_data, on='gene', how='inner')\n",
    "\n",
    "    # calculate fold change and p-value and z-score\n",
    "    combined_data['logFC']= combined_data.apply(\n",
    "        lambda x: calculate_log_fold(x.gene_expression_mutated, x.gene_expression_non_mutated), axis=1)\n",
    "    # calculate wilcox pvalue\n",
    "    combined_data[['pvalue', 'effect_size']] = combined_data.apply(\n",
    "    lambda x: pd.Series(calculate_pvalue_and_effect_size_wilcox_ranksum(x.gene_expression_mutated, x.gene_expression_non_mutated)),\n",
    "    axis=1\n",
    "    )\n",
    "    combined_data['expression_mutated_mean'] = combined_data['gene_expression_mutated'].apply(mean)\n",
    "    combined_data['expression_nonmutated_mean'] = combined_data['gene_expression_non_mutated'].apply(mean)\n",
    "\n",
    "\n",
    "    # calculate adjusted p-value\n",
    "    combined_data['adjusted_pvalue'] = calculate_adjusted_pvalue(combined_data['pvalue'].values)\n",
    "\n",
    "    # Output data to csv \n",
    "    combined_data.drop(columns=['gene_expression_mutated', 'gene_expression_non_mutated'], inplace=True)\n",
    "    \n",
    "    os.makedirs(output_folder, exist_ok=True)\n",
    "    output_filename = f'{output_folder}/{mutated_samples.count()}_{non_mutated_samples.count()}_logfc_pvalue.csv'\n",
    "    \n",
    "    print(f\"outputting data to {output_filename}\")\n",
    "    combined_data.to_csv(output_filename, index=False)\n",
    "\n",
    "    return combined_data, mutated_samples, output_filename\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "ab65d785-dbfe-4336-a4d0-68024188234c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputting data to SPOP/15_102_logfc_pvalue.csv\n"
     ]
    }
   ],
   "source": [
    "volcano_plot_df, individuals_mutated_target_gene, volcano_input_filename = generate_stats_per_gene(\n",
    "    mutation_expression_df_melted, \n",
    "    target_gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e484ffc8-6249-448f-b354-01574bb19717",
   "metadata": {},
   "source": [
    "## generate heatmap of top 100 differentially expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "fe591289-eba9-427a-8563-ada51b200b91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_expression_heatmap(expression_df, volcano_plot_df, n=100, exclude_value=10):\n",
    "    # Get the indices of the top n rows based on absolute values of 'logFC'\n",
    "    # get the genes most differentially expressed (high log FC values)\n",
    "    # if we exclude -exclude_value, exclude_value values\n",
    "    # top_n_indices = volcano_plot_df[\n",
    "    #     (volcano_plot_df['logFC'].abs() != exclude_value)\n",
    "    # ]['logFC'].abs().nlargest(n).index\n",
    "    \n",
    "    # top_n_rows = volcano_plot_df.loc[top_n_indices].set_index('gene')\n",
    "\n",
    "    top_n_indices = volcano_plot_df['pvalue'].nsmallest(n).index\n",
    "    \n",
    "    top_n_rows = volcano_plot_df.loc[top_n_indices].set_index('gene')\n",
    "\n",
    "    # go back to the expression df and make a heatmap of the top n genes\n",
    "    expression_df_heatmap = expression_df.loc[top_n_rows.index]\n",
    "    \n",
    "    return expression_df_heatmap\n",
    "\n",
    "heatmap_data = generate_expression_heatmap(expression_df, volcano_plot_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "e51e4364-9eec-4ef8-ba66-16410c606ac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SPOP/sample_mutation_status.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_mutated_status(expression_df_heatmap, individuals_mutated_target_gene, output_folder=target_gene):\n",
    "    individuals_mutated_target_gene = list(individuals_mutated_target_gene)\n",
    "    \n",
    "    mutated_status = [1 if item in individuals_mutated_target_gene else 0 for item in expression_df_heatmap.columns]\n",
    "\n",
    "    sample_categories_df = pd.DataFrame({\n",
    "        'Sample': expression_df_heatmap.columns,\n",
    "        'Mutation Status': mutated_status\n",
    "    })\n",
    "\n",
    "    output_filename = f'{output_folder}/sample_mutation_status.csv'\n",
    "    print(output_filename)\n",
    "    sample_categories_df.to_csv(output_filename,  index=False)\n",
    "    return sample_categories_df\n",
    "\n",
    "mutated_status_df = get_mutated_status(heatmap_data, individuals_mutated_target_gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "9af3d157-2ce3-4456-a6d9-a7ecbbaab5c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RESPOND_10100218</th>\n",
       "      <th>RESPOND_10100291</th>\n",
       "      <th>RESPOND_10100412</th>\n",
       "      <th>RESPOND_10100478</th>\n",
       "      <th>RESPOND_10100596</th>\n",
       "      <th>RESPOND_10100615</th>\n",
       "      <th>RESPOND_10100801</th>\n",
       "      <th>RESPOND_10100884</th>\n",
       "      <th>RESPOND_10100899</th>\n",
       "      <th>RESPOND_10100952</th>\n",
       "      <th>...</th>\n",
       "      <th>RESPOND_80100242</th>\n",
       "      <th>RESPOND_80100313</th>\n",
       "      <th>RESPOND_80100345</th>\n",
       "      <th>RESPOND_80100411</th>\n",
       "      <th>RESPOND_80100526</th>\n",
       "      <th>RESPOND_80100556</th>\n",
       "      <th>RESPOND_80100590</th>\n",
       "      <th>RESPOND_81100031</th>\n",
       "      <th>RESPOND_40100842</th>\n",
       "      <th>RESPOND_80100259</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARHGEF16</th>\n",
       "      <td>5.554589</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>5.754888</td>\n",
       "      <td>4.321928</td>\n",
       "      <td>6.672425</td>\n",
       "      <td>4.087463</td>\n",
       "      <td>5.977280</td>\n",
       "      <td>6.614710</td>\n",
       "      <td>4.807355</td>\n",
       "      <td>4.700440</td>\n",
       "      <td>...</td>\n",
       "      <td>6.768184</td>\n",
       "      <td>5.754888</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>7.179909</td>\n",
       "      <td>6.794416</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>5.169925</td>\n",
       "      <td>6.285402</td>\n",
       "      <td>5.807355</td>\n",
       "      <td>6.303781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMG20B</th>\n",
       "      <td>10.707359</td>\n",
       "      <td>8.781360</td>\n",
       "      <td>9.667112</td>\n",
       "      <td>8.426265</td>\n",
       "      <td>9.915879</td>\n",
       "      <td>9.308339</td>\n",
       "      <td>10.314017</td>\n",
       "      <td>8.044394</td>\n",
       "      <td>8.361944</td>\n",
       "      <td>8.299208</td>\n",
       "      <td>...</td>\n",
       "      <td>10.392317</td>\n",
       "      <td>10.890264</td>\n",
       "      <td>8.499846</td>\n",
       "      <td>11.106563</td>\n",
       "      <td>9.874981</td>\n",
       "      <td>9.154818</td>\n",
       "      <td>9.491853</td>\n",
       "      <td>9.292322</td>\n",
       "      <td>9.296916</td>\n",
       "      <td>10.213104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLXNB2</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>9.103288</td>\n",
       "      <td>9.623881</td>\n",
       "      <td>9.388017</td>\n",
       "      <td>9.977280</td>\n",
       "      <td>8.918863</td>\n",
       "      <td>9.262095</td>\n",
       "      <td>9.074141</td>\n",
       "      <td>8.686501</td>\n",
       "      <td>9.169925</td>\n",
       "      <td>...</td>\n",
       "      <td>9.611025</td>\n",
       "      <td>10.292322</td>\n",
       "      <td>8.810572</td>\n",
       "      <td>10.650154</td>\n",
       "      <td>10.079485</td>\n",
       "      <td>8.625709</td>\n",
       "      <td>9.337622</td>\n",
       "      <td>9.612868</td>\n",
       "      <td>9.317413</td>\n",
       "      <td>9.569856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WFIKKN1</th>\n",
       "      <td>2.321928</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.857649</td>\n",
       "      <td>...</td>\n",
       "      <td>3.906891</td>\n",
       "      <td>3.321928</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.459432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MC1R</th>\n",
       "      <td>4.459432</td>\n",
       "      <td>7.330917</td>\n",
       "      <td>6.228819</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>6.266787</td>\n",
       "      <td>4.523562</td>\n",
       "      <td>4.906891</td>\n",
       "      <td>6.228819</td>\n",
       "      <td>5.754888</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>...</td>\n",
       "      <td>6.629357</td>\n",
       "      <td>6.523562</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>7.965784</td>\n",
       "      <td>7.483816</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>7.523562</td>\n",
       "      <td>6.392317</td>\n",
       "      <td>6.209453</td>\n",
       "      <td>7.475733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBD3</th>\n",
       "      <td>6.599913</td>\n",
       "      <td>6.189825</td>\n",
       "      <td>7.339850</td>\n",
       "      <td>6.108524</td>\n",
       "      <td>8.022368</td>\n",
       "      <td>6.321928</td>\n",
       "      <td>6.870365</td>\n",
       "      <td>7.651052</td>\n",
       "      <td>5.357552</td>\n",
       "      <td>6.247928</td>\n",
       "      <td>...</td>\n",
       "      <td>7.033423</td>\n",
       "      <td>7.434628</td>\n",
       "      <td>5.169925</td>\n",
       "      <td>8.370687</td>\n",
       "      <td>8.118941</td>\n",
       "      <td>4.954196</td>\n",
       "      <td>7.011227</td>\n",
       "      <td>7.977280</td>\n",
       "      <td>6.700440</td>\n",
       "      <td>7.033423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYB561D1</th>\n",
       "      <td>3.459432</td>\n",
       "      <td>5.087463</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>4.087463</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>2.807355</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>4.321928</td>\n",
       "      <td>4.087463</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.906891</td>\n",
       "      <td>3.906891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GUCY1B1</th>\n",
       "      <td>3.169925</td>\n",
       "      <td>3.321928</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>2.807355</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.906891</td>\n",
       "      <td>4.247928</td>\n",
       "      <td>4.169925</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>2.807355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF1</th>\n",
       "      <td>2.321928</td>\n",
       "      <td>0.371710</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.931392</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044200</td>\n",
       "      <td>0.449118</td>\n",
       "      <td>0.203308</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.102749</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.852043</td>\n",
       "      <td>0.879822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042794</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.078879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEGF6</th>\n",
       "      <td>3.584963</td>\n",
       "      <td>6.266787</td>\n",
       "      <td>5.672425</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>6.375039</td>\n",
       "      <td>4.754888</td>\n",
       "      <td>4.906891</td>\n",
       "      <td>6.087463</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>5.209453</td>\n",
       "      <td>...</td>\n",
       "      <td>6.894818</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>6.087463</td>\n",
       "      <td>5.247928</td>\n",
       "      <td>5.129283</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.087463</td>\n",
       "      <td>7.022368</td>\n",
       "      <td>6.022368</td>\n",
       "      <td>5.285402</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 164 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          RESPOND_10100218  RESPOND_10100291  RESPOND_10100412  \\\n",
       "gene                                                             \n",
       "ARHGEF16          5.554589          5.426265          5.754888   \n",
       "HMG20B           10.707359          8.781360          9.667112   \n",
       "PLXNB2           10.000000          9.103288          9.623881   \n",
       "WFIKKN1           2.321928          3.459432          3.169925   \n",
       "MC1R              4.459432          7.330917          6.228819   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              6.599913          6.189825          7.339850   \n",
       "CYB561D1          3.459432          5.087463          4.000000   \n",
       "GUCY1B1           3.169925          3.321928          3.000000   \n",
       "KLF1              2.321928          0.371710          0.000000   \n",
       "MEGF6             3.584963          6.266787          5.672425   \n",
       "\n",
       "          RESPOND_10100478  RESPOND_10100596  RESPOND_10100615  \\\n",
       "gene                                                             \n",
       "ARHGEF16          4.321928          6.672425          4.087463   \n",
       "HMG20B            8.426265          9.915879          9.308339   \n",
       "PLXNB2            9.388017          9.977280          8.918863   \n",
       "WFIKKN1           2.321928          3.584963          1.584963   \n",
       "MC1R              5.000000          6.266787          4.523562   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              6.108524          8.022368          6.321928   \n",
       "CYB561D1          3.459432          4.087463          3.700440   \n",
       "GUCY1B1           5.426265          2.807355          3.459432   \n",
       "KLF1              0.000000          0.931392          0.000000   \n",
       "MEGF6             4.584963          6.375039          4.754888   \n",
       "\n",
       "          RESPOND_10100801  RESPOND_10100884  RESPOND_10100899  \\\n",
       "gene                                                             \n",
       "ARHGEF16          5.977280          6.614710          4.807355   \n",
       "HMG20B           10.314017          8.044394          8.361944   \n",
       "PLXNB2            9.262095          9.074141          8.686501   \n",
       "WFIKKN1           3.000000          3.000000          2.000000   \n",
       "MC1R              4.906891          6.228819          5.754888   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              6.870365          7.651052          5.357552   \n",
       "CYB561D1          3.459432          3.807355          3.169925   \n",
       "GUCY1B1           3.906891          4.247928          4.169925   \n",
       "KLF1              0.044200          0.449118          0.203308   \n",
       "MEGF6             4.906891          6.087463          4.584963   \n",
       "\n",
       "          RESPOND_10100952  ...  RESPOND_80100242  RESPOND_80100313  \\\n",
       "gene                        ...                                       \n",
       "ARHGEF16          4.700440  ...          6.768184          5.754888   \n",
       "HMG20B            8.299208  ...         10.392317         10.890264   \n",
       "PLXNB2            9.169925  ...          9.611025         10.292322   \n",
       "WFIKKN1           0.857649  ...          3.906891          3.321928   \n",
       "MC1R              4.392317  ...          6.629357          6.523562   \n",
       "...                    ...  ...               ...               ...   \n",
       "MBD3              6.247928  ...          7.033423          7.434628   \n",
       "CYB561D1          2.584963  ...          2.584963          4.392317   \n",
       "GUCY1B1           3.584963  ...          2.584963          1.584963   \n",
       "KLF1              0.000000  ...          0.102749          1.000000   \n",
       "MEGF6             5.209453  ...          6.894818          5.000000   \n",
       "\n",
       "          RESPOND_80100345  RESPOND_80100411  RESPOND_80100526  \\\n",
       "gene                                                             \n",
       "ARHGEF16          3.459432          7.179909          6.794416   \n",
       "HMG20B            8.499846         11.106563          9.874981   \n",
       "PLXNB2            8.810572         10.650154         10.079485   \n",
       "WFIKKN1           1.000000          5.426265          3.459432   \n",
       "MC1R              4.584963          7.965784          7.483816   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              5.169925          8.370687          8.118941   \n",
       "CYB561D1          2.807355          4.392317          4.321928   \n",
       "GUCY1B1           4.392317          2.000000          2.584963   \n",
       "KLF1              0.000000          0.852043          0.879822   \n",
       "MEGF6             6.087463          5.247928          5.129283   \n",
       "\n",
       "          RESPOND_80100556  RESPOND_80100590  RESPOND_81100031  \\\n",
       "gene                                                             \n",
       "ARHGEF16          3.700440          5.169925          6.285402   \n",
       "HMG20B            9.154818          9.491853          9.292322   \n",
       "PLXNB2            8.625709          9.337622          9.612868   \n",
       "WFIKKN1           0.000000          2.000000          3.584963   \n",
       "MC1R              2.321928          7.523562          6.392317   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              4.954196          7.011227          7.977280   \n",
       "CYB561D1          4.087463          3.807355          3.459432   \n",
       "GUCY1B1           3.584963          3.169925          3.000000   \n",
       "KLF1              0.000000          0.042794          1.000000   \n",
       "MEGF6             5.000000          5.087463          7.022368   \n",
       "\n",
       "          RESPOND_40100842  RESPOND_80100259  \n",
       "gene                                          \n",
       "ARHGEF16          5.807355          6.303781  \n",
       "HMG20B            9.296916         10.213104  \n",
       "PLXNB2            9.317413          9.569856  \n",
       "WFIKKN1           3.459432          3.459432  \n",
       "MC1R              6.209453          7.475733  \n",
       "...                    ...               ...  \n",
       "MBD3              6.700440          7.033423  \n",
       "CYB561D1          3.906891          3.906891  \n",
       "GUCY1B1           3.807355          2.807355  \n",
       "KLF1              0.000000          0.078879  \n",
       "MEGF6             6.022368          5.285402  \n",
       "\n",
       "[100 rows x 164 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "heatmap_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "74d85b53-9bb0-478f-a08d-d7d0f8186c41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row Silhouette Score: 0.24124544217953459\n",
      "Column Silhouette Score: 0.06585128389732592\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if not hasattr(array, \"sparse\") and array.dtypes.apply(is_sparse).any():\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if not hasattr(array, \"sparse\") and array.dtypes.apply(is_sparse).any():\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):\n"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.cluster.hierarchy import fcluster, linkage\n",
    "from sklearn.metrics import silhouette_score\n",
    "\n",
    "def hierarchical_clustering(expression_df_heatmap, row_threshold=7, col_threshold=7, output_folder=target_gene):\n",
    "    # Cluster the rows and columns using hierarchical clustering\n",
    "    row_linkage = linkage(expression_df_heatmap, method='ward')\n",
    "    col_linkage = linkage(expression_df_heatmap.T, method='ward')\n",
    "\n",
    "    # Assign cluster labels using cluster\n",
    "    row_clusters = fcluster(row_linkage, t=row_threshold, criterion='maxclust')\n",
    "    col_clusters = fcluster(col_linkage, t=col_threshold, criterion='maxclust')\n",
    "\n",
    "    # Calculate silhouette scores\n",
    "    row_silhouette_score = silhouette_score(expression_df_heatmap, row_clusters)\n",
    "    col_silhouette_score = silhouette_score(expression_df_heatmap.T, col_clusters)\n",
    "\n",
    "    print(f\"Row Silhouette Score: {row_silhouette_score}\")\n",
    "    print(f\"Column Silhouette Score: {col_silhouette_score}\")\n",
    "\n",
    "    # Create dictionaries to store cluster information\n",
    "    row_cluster_info = {f\"Cluster {cluster}\": expression_df_heatmap.index[row_clusters == cluster] for cluster in np.unique(row_clusters)}\n",
    "    col_cluster_info = {f\"Cluster {cluster}\": expression_df_heatmap.columns[col_clusters == cluster] for cluster in np.unique(col_clusters)}\n",
    "\n",
    "    row_cluster_file = os.path.join(output_folder, 'row_clusters.txt')\n",
    "    with open(row_cluster_file, 'w') as file:\n",
    "        file.write(f\"Row Silhouette Score: {row_silhouette_score}\\n\\n\")\n",
    "        for cluster, items in row_cluster_info.items():\n",
    "            file.write(f\"{cluster}:\\n\")\n",
    "            file.write(f\"{', '.join(items)}\\n\\n\")\n",
    "\n",
    "    # Save column cluster information to a text file\n",
    "    col_cluster_file = os.path.join(output_folder, 'col_clusters.txt')\n",
    "    with open(col_cluster_file, 'w') as file:\n",
    "        file.write(f\"Column Silhouette Score: {col_silhouette_score}\\n\\n\")\n",
    "        for cluster, items in col_cluster_info.items():\n",
    "            file.write(f\"{cluster}:\\n\")\n",
    "            file.write(f\"{', '.join(items)}\\n\\n\")\n",
    "    \n",
    "    return row_linkage, col_linkage, row_cluster_info, col_cluster_info\n",
    "\n",
    "\n",
    "row_linkage, col_linkage, row_clusters, col_clusters = hierarchical_clustering(heatmap_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "07391585-1787-4d24-85c9-8c597915f80c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def create_clustered_heatmap_and_save(expression_df_heatmap, \n",
    "                                      row_linkage, \n",
    "                                      col_linkage, \n",
    "                                      sample_mutation_df, \n",
    "                                      output_folder=target_gene):\n",
    "    # Group labels and colors for color bar\n",
    "    type_map = {1: 'red', 0: 'yellow'}\n",
    "\n",
    "    sns.set(font_scale=1.0) \n",
    "\n",
    "    # Create a clustered heatmap\n",
    "    clustered_df = sns.clustermap(\n",
    "        expression_df_heatmap,\n",
    "        row_linkage=row_linkage,\n",
    "        col_linkage=col_linkage,\n",
    "        cmap='viridis',\n",
    "        annot=False,\n",
    "        fmt=\".1f\",\n",
    "        linewidths=.5,\n",
    "        col_colors=sample_mutation_df.set_index('Sample')['Mutation Status'].map(type_map),\n",
    "        z_score=0,\n",
    "        cbar_kws={\"shrink\": 0.7, \"aspect\": 30}  # Adjust color bar size\n",
    "    )\n",
    "\n",
    "    #Set the size of the overall figure\n",
    "    clustered_df.fig.set_size_inches(15, len(expression_df_heatmap) * 0.2)  \n",
    "\n",
    "    output_filename = f'{output_folder}/heatmap.png'\n",
    "\n",
    "    clustered_df.ax_heatmap.set_yticklabels(clustered_df.ax_heatmap.get_yticklabels(), rotation=0)\n",
    "    \n",
    "    # Save the plot to a file\n",
    "    clustered_df.savefig(output_filename, bbox_inches='tight')\n",
    "\n",
    "    # Close the plot to prevent displaying it in the notebook (optional)\n",
    "    plt.close()\n",
    "\n",
    "    return clustered_df\n",
    "\n",
    "clustered_heatmap = create_clustered_heatmap_and_save(heatmap_data, \n",
    "                                                      row_linkage, \n",
    "                                                      col_linkage,\n",
    "                                                      mutated_status_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "56c644ef-b34e-4343-b27e-e281a2be9aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram\n",
    "\n",
    "## add more args to the function \n",
    "\n",
    "def plot_and_save_dendrograms(row_linkage, col_linkage, expression_df_heatmap, output_folder=target_gene):\n",
    "    # Plot the row dendrogram\n",
    "    plt.figure(figsize=(12, 10))\n",
    "    # play around with fig sizes \n",
    "    gene_plot = dendrogram(row_linkage, \n",
    "                           labels=expression_df_heatmap.index, \n",
    "                           truncate_mode='lastp', \n",
    "                           p=int(0.75 * len(row_linkage)), show_leaf_counts=True)\n",
    "    plt.title('Row Dendrogram')\n",
    "    \n",
    "    # Save the row dendrogram plot to a file\n",
    "    row_dendrogram_filename = f'{output_folder}/row_dendrogram.png'\n",
    "    plt.savefig(row_dendrogram_filename)\n",
    "    plt.close()\n",
    "\n",
    "    # Plot the column dendrogram\n",
    "    # play around with fig sizes \n",
    "    plt.figure(figsize=(12, 10))\n",
    "    sample_plot = dendrogram(col_linkage, labels=expression_df_heatmap.columns, orientation='top', \n",
    "                             truncate_mode='lastp', p=int(0.75 * len(row_linkage)), show_leaf_counts=True)\n",
    "    plt.title('Column Dendrogram')\n",
    "    \n",
    "    # Save the column dendrogram plot to a file\n",
    "    col_dendrogram_filename = f'{output_folder}/col_dendrogram.png'\n",
    "    plt.savefig(col_dendrogram_filename)\n",
    "    plt.close()\n",
    "\n",
    "    return gene_plot, sample_plot\n",
    "\n",
    "row_dendrogram, col_dendrogram = plot_and_save_dendrograms(row_linkage, \n",
    "                                                           col_linkage, \n",
    "                                                           heatmap_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "0e7b9c4a-c204-40a4-af6f-a80cbb4a0594",
   "metadata": {},
   "outputs": [],
   "source": [
    "def histogram_of_column_and_save(df, column, output_folder=target_gene):\n",
    "    # Plot the distribution of p-values\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.hist(df[column], bins=30, color='blue', edgecolor='black')\n",
    "    plt.title(f'Distribution of {column}')\n",
    "    plt.xlabel('P-values')\n",
    "    plt.ylabel('Frequency')\n",
    "\n",
    "    output_filename = f'{output_folder}/histogram_{column}.png'\n",
    "    # Save the histogram plot to a file\n",
    "    plt.savefig(output_filename)\n",
    "    plt.close()\n",
    "\n",
    "histogram_of_column_and_save(volcano_plot_df, 'pvalue')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "0088ccb1-ed8b-4faa-a98f-4f69963bc1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def volcano_plot(input_file_path, yaxis, \n",
    "                 xaxis, \n",
    "                 significance_threshold=0.05, \n",
    "                 logfold_positive_threshold=2, \n",
    "                 logfold_negative_threshold=-2,\n",
    "                 output_folder=target_gene):\n",
    "    df = pd.read_csv(input_file_path)\n",
    "\n",
    "    # Apply -log10 transformation to the p-value\n",
    "    df['-log10_pvalue'] = -np.log10(df[yaxis])\n",
    "\n",
    "    # Highlight significant points with large log-fold changes\n",
    "    significant_genes_positive = df[(df[yaxis] < significance_threshold) & (df['logFC'] > logfold_positive_threshold)]\n",
    "    significant_genes_negative = df[(df[yaxis] < significance_threshold) & (df['logFC'] < logfold_negative_threshold)]\n",
    "\n",
    "    # Clip outliers in 'logFC' column for the plot\n",
    "    df['logFC'] = df['logFC'].clip(lower=-10, upper=10)\n",
    "\n",
    "    plt.scatter(x=df[xaxis], y=df['-log10_pvalue'], s=1, label='All Genes', alpha=0.5)\n",
    "    plt.scatter(x=significant_genes_positive[xaxis], y=significant_genes_positive['-log10_pvalue'], s=10, c='red', marker='^', label='Significant Genes (Positive LogFC)')\n",
    "    plt.scatter(x=significant_genes_negative[xaxis], y=significant_genes_negative['-log10_pvalue'], s=10, c='blue', marker='v', label='Significant Genes (Negative LogFC)')\n",
    "\n",
    "    plt.xlabel(xaxis)\n",
    "    plt.ylabel(f'-log10({yaxis})')\n",
    "    plt.title('Volcano Plot')\n",
    "    plt.axhline(-np.log10(significance_threshold), color='gray', linestyle='--', label=f'Significance Threshold ({significance_threshold})')\n",
    "    plt.axvline(logfold_positive_threshold, color='gray', linestyle='--', label=f'Positive Log-Fold Change Threshold ({logfold_positive_threshold})')\n",
    "    plt.axvline(logfold_negative_threshold, color='gray', linestyle='--', label=f'Negative Log-Fold Change Threshold ({logfold_negative_threshold})')\n",
    "    \n",
    "     # Create a separate legend outside the plot\n",
    "    fig = plt.gcf()\n",
    "    handles, labels = plt.gca().get_legend_handles_labels()\n",
    "    fig.legend(handles, labels, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    \n",
    "    \n",
    "    output_filename = f'{output_folder}/volcano_plot.png'\n",
    "    plt.savefig(output_filename, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    significant_genes_positive.to_csv(f'{output_folder}/signif_genes_positive.csv')\n",
    "    significant_genes_negative.to_csv(f'{output_folder}/signif_genes_negative.csv')\n",
    "\n",
    "    return significant_genes_positive, significant_genes_negative\n",
    "\n",
    "significant_genes_positive, significant_genes_negative  = volcano_plot(volcano_input_filename, 'pvalue', 'logFC')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "7f2a377a-206b-4256-a9ed-39d78bfe3bf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def create_gene_expression_boxplot(expression_df, significant_genes_df, mutated_status_df, output_folder=target_gene, positive=1 ):\n",
    "    # Filter for significant genes\n",
    "    filtered_expression_df_genes_of_interest = expression_df.loc[significant_genes_df.gene]\n",
    "\n",
    "    # Unstack the expression DataFrame\n",
    "    unstack_expression = filtered_expression_df_genes_of_interest.unstack().reset_index()\n",
    "    unstack_expression.columns = ['Sample', 'Gene', 'Expression']\n",
    "\n",
    "    # Merge the expression matrix with the mutation information\n",
    "    boxplot_df = pd.merge(unstack_expression, mutated_status_df, how='inner', on='Sample')\n",
    "\n",
    "    # Rename columns for clarity\n",
    "    boxplot_df.columns = ['Sample', 'Gene', 'Expression', 'Target Gene Mutation Status']\n",
    "\n",
    "    # Set up the plot\n",
    "    plt.figure(figsize=(12, 8))\n",
    "\n",
    "    # Create a box plot with 'hue' for each gene and 'dodge' for Mutation Status\n",
    "    sns.boxplot(x='Gene', y='Expression', data=boxplot_df, hue='Target Gene Mutation Status', dodge=True)\n",
    "\n",
    "    # Customize the plot\n",
    "    plt.xlabel('Gene')\n",
    "    plt.ylabel('Gene Expression')\n",
    "    plt.title('Box Plots of Gene Expression for Mutated and Non-Mutated Individuals')\n",
    "\n",
    "    # Rotate x-axis labels for better readability\n",
    "    plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "    if positive == 1:\n",
    "        output_filename = f'{output_folder}/positive_genes_expression_boxplot.png'\n",
    "    else:\n",
    "        output_filename = f'{output_folder}/negative_genes_expression_boxplot.png'\n",
    "    plt.savefig(output_filename)\n",
    "    plt.close()\n",
    "\n",
    "create_gene_expression_boxplot(expression_df, significant_genes_positive, mutated_status_df, positive=1)\n",
    "create_gene_expression_boxplot(expression_df, significant_genes_negative, mutated_status_df, positive=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "765c7969-47bd-4c9f-8055-958e5fb1ca6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5adbbf0a-9530-4fc4-91b2-1c3214236512",
   "metadata": {},
   "source": [
    "# exploring mutation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4966b00d-1a74-4586-a108-2ca8658a63d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'gsea' from '/Users/meltemtutar/Documents/Huang/Respond/gsea.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "from importlib import reload\n",
    "import plot, utils, data_load, clustering, constants, gsea \n",
    "import warnings\n",
    "\n",
    "# Suppress FutureWarnings from multiple modules\n",
    "modules_to_ignore = (\"seaborn\", \"sklearn\")\n",
    "\n",
    "for module in modules_to_ignore:\n",
    "    warnings.filterwarnings(\"ignore\", category=FutureWarning, module=module)\n",
    "\n",
    "reload(clustering)\n",
    "reload(utils)\n",
    "reload(data_load)\n",
    "reload(plot)\n",
    "reload(constants)\n",
    "reload(gsea)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ee347d44-54ef-49e9-bfda-3a44612ffeda",
   "metadata": {},
   "outputs": [],
   "source": [
    "#hugo_symbol means gene, Tumor_Sample_Barcode is the sampleid\n",
    "columns = [\"Hugo_Symbol\", \"Tumor_Sample_Barcode\", \"Variant_Classification\", \"Variant_Type\", \"Chromosome\", \"Start_Position\", \"End_Position\",\n",
    "           \"Description\", \"Protein_Change\"]\n",
    "maf_df = data_load.load_maf_data(file_path=constants.maf_file_path, columns=columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f925014-b6b7-4bda-ab30-469e13ceda5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>sample</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>Description</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6264</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49038778</td>\n",
       "      <td>49038778</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20100780</td>\n",
       "      <td>p.R2860C</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6265</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49041176</td>\n",
       "      <td>49041176</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>RESPOND_60100060</td>\n",
       "      <td>p.Y2199fs</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6266</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49031631</td>\n",
       "      <td>49031668</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>RESPOND_60100201</td>\n",
       "      <td>p.P4346fs</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6267</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49043078</td>\n",
       "      <td>49043079</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>RESPOND_60100201</td>\n",
       "      <td>p.E1881fs</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49041009</td>\n",
       "      <td>49041009</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_60100883</td>\n",
       "      <td>p.N2254S</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6269</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49051204</td>\n",
       "      <td>49051204</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_61100223</td>\n",
       "      <td>p.Q827*</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49048052</td>\n",
       "      <td>49048052</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_61100364</td>\n",
       "      <td>p.C1383W</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13972</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49040614</td>\n",
       "      <td>49040614</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100563</td>\n",
       "      <td>p.R2386W</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14394</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49049713</td>\n",
       "      <td>49049713</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20107556</td>\n",
       "      <td>p.R1292H</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15721</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49039268</td>\n",
       "      <td>49039268</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_60100690</td>\n",
       "      <td>p.R2774W</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15812</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49040403</td>\n",
       "      <td>49040403</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20104864</td>\n",
       "      <td>p.R2456H</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15877</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49038220</td>\n",
       "      <td>49038220</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_40102910</td>\n",
       "      <td>p.E3046K</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15944</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49041265</td>\n",
       "      <td>49041265</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_60100300</td>\n",
       "      <td>p.A2169T</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16004</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49038777</td>\n",
       "      <td>49038777</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70101764</td>\n",
       "      <td>p.R2860H</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16027</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49041375</td>\n",
       "      <td>49041375</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10101427</td>\n",
       "      <td>p.P2132L</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16046</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49049761</td>\n",
       "      <td>49049761</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_40100784</td>\n",
       "      <td>p.G1276E</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16060</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49038867</td>\n",
       "      <td>49038867</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100229</td>\n",
       "      <td>p.R2830Q</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16122</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49042630</td>\n",
       "      <td>49042630</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20101031</td>\n",
       "      <td>p.G1933D</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16193</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49044277</td>\n",
       "      <td>49044277</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100042</td>\n",
       "      <td>p.R1704Q</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16211</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49026947</td>\n",
       "      <td>49026947</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100302</td>\n",
       "      <td>p.R5007W</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16236</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49028137</td>\n",
       "      <td>49028137</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100812</td>\n",
       "      <td>p.L4796P</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16237</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49031255</td>\n",
       "      <td>49031255</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100812</td>\n",
       "      <td>p.R4484*</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16888</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49040118</td>\n",
       "      <td>49040118</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20104136</td>\n",
       "      <td>p.V2551G</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17570</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49042589</td>\n",
       "      <td>49042589</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100286</td>\n",
       "      <td>p.L1947F</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18097</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49039837</td>\n",
       "      <td>49039837</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10100029</td>\n",
       "      <td>p.R2645*</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18429</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49038305</td>\n",
       "      <td>49038306</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>RESPOND_10100462</td>\n",
       "      <td>p.D3017fs</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18609</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49030729</td>\n",
       "      <td>49030729</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10100562</td>\n",
       "      <td>p.A4571T</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18750</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49026790</td>\n",
       "      <td>49026790</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10100653</td>\n",
       "      <td>p.H5059P</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18751</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49046080</td>\n",
       "      <td>49046080</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10100653</td>\n",
       "      <td>p.V1560M</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19482</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49022732</td>\n",
       "      <td>49022734</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>RESPOND_20100111</td>\n",
       "      <td>p.5398_5399NV&gt;K</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19646</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49044277</td>\n",
       "      <td>49044277</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20100299</td>\n",
       "      <td>p.R1704Q</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20974</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49038096</td>\n",
       "      <td>49038096</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20100886</td>\n",
       "      <td>p.R3087Q</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23066</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49031843</td>\n",
       "      <td>49031843</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20102454</td>\n",
       "      <td>p.R4288W</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23067</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49048658</td>\n",
       "      <td>49048658</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20102454</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23363</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49049884</td>\n",
       "      <td>49049884</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>RESPOND_20105623</td>\n",
       "      <td>p.G1235fs</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23545</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49022108</td>\n",
       "      <td>49022108</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20106228</td>\n",
       "      <td>p.V5486M</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23976</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49033586</td>\n",
       "      <td>49033586</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_40100334</td>\n",
       "      <td>p.R3707*</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23977</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49040199</td>\n",
       "      <td>49040199</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_40100334</td>\n",
       "      <td>p.T2524M</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24488</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49051305</td>\n",
       "      <td>49051318</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>RESPOND_40100837</td>\n",
       "      <td>p.S789fs</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25474</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49041150</td>\n",
       "      <td>49041150</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_50101079</td>\n",
       "      <td>p.A2207V</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26122</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49039975</td>\n",
       "      <td>49039975</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_61100192</td>\n",
       "      <td>p.G2599R</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26360</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49042767</td>\n",
       "      <td>49042767</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100230</td>\n",
       "      <td>p.T1919M</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27373</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49037710</td>\n",
       "      <td>49037710</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10100085</td>\n",
       "      <td>p.E3216K</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27375</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49029075</td>\n",
       "      <td>49029075</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10100462</td>\n",
       "      <td>p.R4746Q</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27398</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49032233</td>\n",
       "      <td>49032233</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10101119</td>\n",
       "      <td>p.P4158S</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27435</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49024702</td>\n",
       "      <td>49024702</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20100631</td>\n",
       "      <td>p.G5310R</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27484</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49046260</td>\n",
       "      <td>49046260</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20101716</td>\n",
       "      <td>p.R1528Q</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27497</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49031984</td>\n",
       "      <td>49031984</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20101824</td>\n",
       "      <td>p.P4241S</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27515</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49039759</td>\n",
       "      <td>49039759</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20103012</td>\n",
       "      <td>p.G2671S</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27613</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49031662</td>\n",
       "      <td>49031662</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100230</td>\n",
       "      <td>p.G4348E</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27614</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49039866</td>\n",
       "      <td>49039866</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_70100230</td>\n",
       "      <td>p.R2635Q</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27680</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49032334</td>\n",
       "      <td>49032334</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_10100829</td>\n",
       "      <td>p.S4124C</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27840</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49041487</td>\n",
       "      <td>49041487</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20101598</td>\n",
       "      <td>p.R2095C</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27997</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49050032</td>\n",
       "      <td>49050032</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_61100037</td>\n",
       "      <td>p.E1186*</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28858</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49022108</td>\n",
       "      <td>49022108</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20101417</td>\n",
       "      <td>p.V5486M</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28913</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49041121</td>\n",
       "      <td>49041121</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_61100383</td>\n",
       "      <td>p.R2217C</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28943</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49026889</td>\n",
       "      <td>49026889</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20101598</td>\n",
       "      <td>p.P5026L</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28944</th>\n",
       "      <td>KMT2D</td>\n",
       "      <td>chr12</td>\n",
       "      <td>49037919</td>\n",
       "      <td>49037919</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>RESPOND_20101598</td>\n",
       "      <td>p.A3146V</td>\n",
       "      <td>lysine methyltransferase 2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene Chromosome  Start_Position  End_Position Variant_Classification  \\\n",
       "6264   KMT2D      chr12        49038778      49038778      Missense_Mutation   \n",
       "6265   KMT2D      chr12        49041176      49041176        Frame_Shift_Del   \n",
       "6266   KMT2D      chr12        49031631      49031668        Frame_Shift_Del   \n",
       "6267   KMT2D      chr12        49043078      49043079        Frame_Shift_Del   \n",
       "6268   KMT2D      chr12        49041009      49041009      Missense_Mutation   \n",
       "6269   KMT2D      chr12        49051204      49051204      Nonsense_Mutation   \n",
       "6270   KMT2D      chr12        49048052      49048052      Missense_Mutation   \n",
       "13972  KMT2D      chr12        49040614      49040614      Missense_Mutation   \n",
       "14394  KMT2D      chr12        49049713      49049713      Missense_Mutation   \n",
       "15721  KMT2D      chr12        49039268      49039268      Missense_Mutation   \n",
       "15812  KMT2D      chr12        49040403      49040403      Missense_Mutation   \n",
       "15877  KMT2D      chr12        49038220      49038220      Missense_Mutation   \n",
       "15944  KMT2D      chr12        49041265      49041265      Missense_Mutation   \n",
       "16004  KMT2D      chr12        49038777      49038777      Missense_Mutation   \n",
       "16027  KMT2D      chr12        49041375      49041375      Missense_Mutation   \n",
       "16046  KMT2D      chr12        49049761      49049761      Missense_Mutation   \n",
       "16060  KMT2D      chr12        49038867      49038867      Missense_Mutation   \n",
       "16122  KMT2D      chr12        49042630      49042630      Missense_Mutation   \n",
       "16193  KMT2D      chr12        49044277      49044277      Missense_Mutation   \n",
       "16211  KMT2D      chr12        49026947      49026947      Missense_Mutation   \n",
       "16236  KMT2D      chr12        49028137      49028137      Missense_Mutation   \n",
       "16237  KMT2D      chr12        49031255      49031255      Nonsense_Mutation   \n",
       "16888  KMT2D      chr12        49040118      49040118      Missense_Mutation   \n",
       "17570  KMT2D      chr12        49042589      49042589      Missense_Mutation   \n",
       "18097  KMT2D      chr12        49039837      49039837      Nonsense_Mutation   \n",
       "18429  KMT2D      chr12        49038305      49038306        Frame_Shift_Del   \n",
       "18609  KMT2D      chr12        49030729      49030729      Missense_Mutation   \n",
       "18750  KMT2D      chr12        49026790      49026790      Missense_Mutation   \n",
       "18751  KMT2D      chr12        49046080      49046080      Missense_Mutation   \n",
       "19482  KMT2D      chr12        49022732      49022734           In_Frame_Del   \n",
       "19646  KMT2D      chr12        49044277      49044277      Missense_Mutation   \n",
       "20974  KMT2D      chr12        49038096      49038096      Missense_Mutation   \n",
       "23066  KMT2D      chr12        49031843      49031843      Missense_Mutation   \n",
       "23067  KMT2D      chr12        49048658      49048658            Splice_Site   \n",
       "23363  KMT2D      chr12        49049884      49049884        Frame_Shift_Del   \n",
       "23545  KMT2D      chr12        49022108      49022108      Missense_Mutation   \n",
       "23976  KMT2D      chr12        49033586      49033586      Nonsense_Mutation   \n",
       "23977  KMT2D      chr12        49040199      49040199      Missense_Mutation   \n",
       "24488  KMT2D      chr12        49051305      49051318        Frame_Shift_Del   \n",
       "25474  KMT2D      chr12        49041150      49041150      Missense_Mutation   \n",
       "26122  KMT2D      chr12        49039975      49039975      Missense_Mutation   \n",
       "26360  KMT2D      chr12        49042767      49042767      Missense_Mutation   \n",
       "27373  KMT2D      chr12        49037710      49037710      Missense_Mutation   \n",
       "27375  KMT2D      chr12        49029075      49029075      Missense_Mutation   \n",
       "27398  KMT2D      chr12        49032233      49032233      Missense_Mutation   \n",
       "27435  KMT2D      chr12        49024702      49024702      Missense_Mutation   \n",
       "27484  KMT2D      chr12        49046260      49046260            Splice_Site   \n",
       "27497  KMT2D      chr12        49031984      49031984      Missense_Mutation   \n",
       "27515  KMT2D      chr12        49039759      49039759      Missense_Mutation   \n",
       "27613  KMT2D      chr12        49031662      49031662      Missense_Mutation   \n",
       "27614  KMT2D      chr12        49039866      49039866      Missense_Mutation   \n",
       "27680  KMT2D      chr12        49032334      49032334      Missense_Mutation   \n",
       "27840  KMT2D      chr12        49041487      49041487      Missense_Mutation   \n",
       "27997  KMT2D      chr12        49050032      49050032      Nonsense_Mutation   \n",
       "28858  KMT2D      chr12        49022108      49022108      Missense_Mutation   \n",
       "28913  KMT2D      chr12        49041121      49041121      Missense_Mutation   \n",
       "28943  KMT2D      chr12        49026889      49026889      Missense_Mutation   \n",
       "28944  KMT2D      chr12        49037919      49037919      Missense_Mutation   \n",
       "\n",
       "      Variant_Type            sample   Protein_Change  \\\n",
       "6264           SNP  RESPOND_20100780         p.R2860C   \n",
       "6265           DEL  RESPOND_60100060        p.Y2199fs   \n",
       "6266           DEL  RESPOND_60100201        p.P4346fs   \n",
       "6267           DEL  RESPOND_60100201        p.E1881fs   \n",
       "6268           SNP  RESPOND_60100883         p.N2254S   \n",
       "6269           SNP  RESPOND_61100223          p.Q827*   \n",
       "6270           SNP  RESPOND_61100364         p.C1383W   \n",
       "13972          SNP  RESPOND_70100563         p.R2386W   \n",
       "14394          SNP  RESPOND_20107556         p.R1292H   \n",
       "15721          SNP  RESPOND_60100690         p.R2774W   \n",
       "15812          SNP  RESPOND_20104864         p.R2456H   \n",
       "15877          SNP  RESPOND_40102910         p.E3046K   \n",
       "15944          SNP  RESPOND_60100300         p.A2169T   \n",
       "16004          SNP  RESPOND_70101764         p.R2860H   \n",
       "16027          SNP  RESPOND_10101427         p.P2132L   \n",
       "16046          SNP  RESPOND_40100784         p.G1276E   \n",
       "16060          SNP  RESPOND_70100229         p.R2830Q   \n",
       "16122          SNP  RESPOND_20101031         p.G1933D   \n",
       "16193          SNP  RESPOND_70100042         p.R1704Q   \n",
       "16211          SNP  RESPOND_70100302         p.R5007W   \n",
       "16236          SNP  RESPOND_70100812         p.L4796P   \n",
       "16237          SNP  RESPOND_70100812         p.R4484*   \n",
       "16888          SNP  RESPOND_20104136         p.V2551G   \n",
       "17570          SNP  RESPOND_70100286         p.L1947F   \n",
       "18097          SNP  RESPOND_10100029         p.R2645*   \n",
       "18429          DEL  RESPOND_10100462        p.D3017fs   \n",
       "18609          SNP  RESPOND_10100562         p.A4571T   \n",
       "18750          SNP  RESPOND_10100653         p.H5059P   \n",
       "18751          SNP  RESPOND_10100653         p.V1560M   \n",
       "19482          DEL  RESPOND_20100111  p.5398_5399NV>K   \n",
       "19646          SNP  RESPOND_20100299         p.R1704Q   \n",
       "20974          SNP  RESPOND_20100886         p.R3087Q   \n",
       "23066          SNP  RESPOND_20102454         p.R4288W   \n",
       "23067          SNP  RESPOND_20102454              NaN   \n",
       "23363          DEL  RESPOND_20105623        p.G1235fs   \n",
       "23545          SNP  RESPOND_20106228         p.V5486M   \n",
       "23976          SNP  RESPOND_40100334         p.R3707*   \n",
       "23977          SNP  RESPOND_40100334         p.T2524M   \n",
       "24488          DEL  RESPOND_40100837         p.S789fs   \n",
       "25474          SNP  RESPOND_50101079         p.A2207V   \n",
       "26122          SNP  RESPOND_61100192         p.G2599R   \n",
       "26360          SNP  RESPOND_70100230         p.T1919M   \n",
       "27373          SNP  RESPOND_10100085         p.E3216K   \n",
       "27375          SNP  RESPOND_10100462         p.R4746Q   \n",
       "27398          SNP  RESPOND_10101119         p.P4158S   \n",
       "27435          SNP  RESPOND_20100631         p.G5310R   \n",
       "27484          SNP  RESPOND_20101716         p.R1528Q   \n",
       "27497          SNP  RESPOND_20101824         p.P4241S   \n",
       "27515          SNP  RESPOND_20103012         p.G2671S   \n",
       "27613          SNP  RESPOND_70100230         p.G4348E   \n",
       "27614          SNP  RESPOND_70100230         p.R2635Q   \n",
       "27680          SNP  RESPOND_10100829         p.S4124C   \n",
       "27840          SNP  RESPOND_20101598         p.R2095C   \n",
       "27997          SNP  RESPOND_61100037         p.E1186*   \n",
       "28858          SNP  RESPOND_20101417         p.V5486M   \n",
       "28913          SNP  RESPOND_61100383         p.R2217C   \n",
       "28943          SNP  RESPOND_20101598         p.P5026L   \n",
       "28944          SNP  RESPOND_20101598         p.A3146V   \n",
       "\n",
       "                       Description  mutation  \n",
       "6264   lysine methyltransferase 2D         1  \n",
       "6265   lysine methyltransferase 2D         1  \n",
       "6266   lysine methyltransferase 2D         1  \n",
       "6267   lysine methyltransferase 2D         1  \n",
       "6268   lysine methyltransferase 2D         1  \n",
       "6269   lysine methyltransferase 2D         1  \n",
       "6270   lysine methyltransferase 2D         1  \n",
       "13972  lysine methyltransferase 2D         1  \n",
       "14394  lysine methyltransferase 2D         1  \n",
       "15721  lysine methyltransferase 2D         1  \n",
       "15812  lysine methyltransferase 2D         1  \n",
       "15877  lysine methyltransferase 2D         1  \n",
       "15944  lysine methyltransferase 2D         1  \n",
       "16004  lysine methyltransferase 2D         1  \n",
       "16027  lysine methyltransferase 2D         1  \n",
       "16046  lysine methyltransferase 2D         1  \n",
       "16060  lysine methyltransferase 2D         1  \n",
       "16122  lysine methyltransferase 2D         1  \n",
       "16193  lysine methyltransferase 2D         1  \n",
       "16211  lysine methyltransferase 2D         1  \n",
       "16236  lysine methyltransferase 2D         1  \n",
       "16237  lysine methyltransferase 2D         1  \n",
       "16888  lysine methyltransferase 2D         1  \n",
       "17570  lysine methyltransferase 2D         1  \n",
       "18097  lysine methyltransferase 2D         1  \n",
       "18429  lysine methyltransferase 2D         1  \n",
       "18609  lysine methyltransferase 2D         1  \n",
       "18750  lysine methyltransferase 2D         1  \n",
       "18751  lysine methyltransferase 2D         1  \n",
       "19482  lysine methyltransferase 2D         1  \n",
       "19646  lysine methyltransferase 2D         1  \n",
       "20974  lysine methyltransferase 2D         1  \n",
       "23066  lysine methyltransferase 2D         1  \n",
       "23067  lysine methyltransferase 2D         1  \n",
       "23363  lysine methyltransferase 2D         1  \n",
       "23545  lysine methyltransferase 2D         1  \n",
       "23976  lysine methyltransferase 2D         1  \n",
       "23977  lysine methyltransferase 2D         1  \n",
       "24488  lysine methyltransferase 2D         1  \n",
       "25474  lysine methyltransferase 2D         1  \n",
       "26122  lysine methyltransferase 2D         1  \n",
       "26360  lysine methyltransferase 2D         1  \n",
       "27373  lysine methyltransferase 2D         1  \n",
       "27375  lysine methyltransferase 2D         1  \n",
       "27398  lysine methyltransferase 2D         1  \n",
       "27435  lysine methyltransferase 2D         1  \n",
       "27484  lysine methyltransferase 2D         1  \n",
       "27497  lysine methyltransferase 2D         1  \n",
       "27515  lysine methyltransferase 2D         1  \n",
       "27613  lysine methyltransferase 2D         1  \n",
       "27614  lysine methyltransferase 2D         1  \n",
       "27680  lysine methyltransferase 2D         1  \n",
       "27840  lysine methyltransferase 2D         1  \n",
       "27997  lysine methyltransferase 2D         1  \n",
       "28858  lysine methyltransferase 2D         1  \n",
       "28913  lysine methyltransferase 2D         1  \n",
       "28943  lysine methyltransferase 2D         1  \n",
       "28944  lysine methyltransferase 2D         1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maf_df[maf_df['gene']=='KMT2D']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0e3a081f-19bc-463f-8dfa-050c2f414eec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fraction of rows filtered is 0.2865853658536585\n"
     ]
    }
   ],
   "source": [
    "## load data \n",
    "maf_df = data_load.load_maf_data(file_path=constants.maf_file_path, columns=[\"Hugo_Symbol\", \"Tumor_Sample_Barcode\", \"Variant_Classification\", \"Variant_Type\", \"Chromosome\", \"Start_Position\", \"End_Position\",\n",
    "           \"Description\", \"Protein_Change\"])\n",
    "expression_df = data_load.load_txt_file_into_dataframe(file_path=constants.expression_file_path)\n",
    "expression_df_melted = data_load.reformat_expression_data(df=expression_df)\n",
    "mutation_expression_df_melted = data_load.preprocess_and_combine_mutation_expression(maf_df= maf_df, expression_df = expression_df_melted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d7d788c0-809f-44ca-82e2-75cfbb264899",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputting data to output_data_v9/KMT2D/17_104_logfc_pvalue.csv\n"
     ]
    }
   ],
   "source": [
    "# calculate logfc, pvalue\n",
    "volcano_plot_df, individuals_mutated_target_gene, output_stats_path = utils.generate_stats_per_gene(\n",
    "        express_mut_genes_df=mutation_expression_df_melted, \n",
    "        target_gene='KMT2D',\n",
    "        output_folder=constants.output_folder)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7403855d-5a94-4bf2-b428-39a53fd43b7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "348394     RESPOND_20100780\n",
       "488442     RESPOND_20101031\n",
       "608461     RESPOND_20101598\n",
       "608462     RESPOND_20101598\n",
       "608463     RESPOND_20101598\n",
       "988501     RESPOND_40100784\n",
       "1188518    RESPOND_60100201\n",
       "1188519    RESPOND_60100201\n",
       "1408539    RESPOND_60100883\n",
       "1688567    RESPOND_61100037\n",
       "1788580    RESPOND_61100364\n",
       "1868587    RESPOND_70100042\n",
       "1948596    RESPOND_70100229\n",
       "1968599    RESPOND_70100286\n",
       "2048606    RESPOND_70100563\n",
       "2128616    RESPOND_70100812\n",
       "2128617    RESPOND_70100812\n",
       "Name: sample, dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "individuals_mutated_target_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7c969a9f-0ed9-4b5b-babc-c69cdc7f019d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_expression_df_melted['mutation']\n",
    "samples_cluster_6 = ['RESPOND_20100780', 'RESPOND_20101031', 'RESPOND_20101598',\n",
    "       'RESPOND_40100176', 'RESPOND_40100292', 'RESPOND_40100361',\n",
    "       'RESPOND_40100523', 'RESPOND_50100227', 'RESPOND_60100201',\n",
    "       'RESPOND_60100943', 'RESPOND_60101197', 'RESPOND_61100037',\n",
    "       'RESPOND_70100042', 'RESPOND_70100229', 'RESPOND_70100286',\n",
    "       'RESPOND_70100317', 'RESPOND_70100812', 'RESPOND_70102184',\n",
    "       'RESPOND_80100590']\n",
    "mutated_samples=set(individuals_mutated_target_gene)\n",
    "samples_same_cluster = mutation_expression_df_melted[(mutation_expression_df_melted['mutation'] == 1.0) & \n",
    "(mutation_expression_df_melted['gene'] == 'KMT2D')& \n",
    "(mutation_expression_df_melted['sample'].isin(samples))]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ae37dd5b-124f-4f9d-9748-22ea78df82b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'p.A3146V',\n",
       " 'p.E1186*',\n",
       " 'p.E1881fs',\n",
       " 'p.G1933D',\n",
       " 'p.L1947F',\n",
       " 'p.L4796P',\n",
       " 'p.P4346fs',\n",
       " 'p.P5026L',\n",
       " 'p.R1704Q',\n",
       " 'p.R2095C',\n",
       " 'p.R2830Q',\n",
       " 'p.R2860C',\n",
       " 'p.R4484*'}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Some samples have multiple mutations\n",
    "set(samples_same_cluster['Protein_Change'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2afd2b82-9d1e-4ff4-ba64-8e72b8dd49b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "def load_maf_data_2(file_path):\n",
    "    df = pd.read_csv(file_path, sep='\\t', comment=\"#\")\n",
    "    df['mutation'] = 1\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a8dd0a37-cbcf-4149-bbda-4417e5e67412",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/87/3s003r291fv10lqn34cwt0c80000gn/T/ipykernel_8386/836280929.py:4: DtypeWarning: Columns (13,91,93,98,135,147,174,176,184,186,188,190,210,211,213,221,224,225) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(file_path, sep='\\t', comment=\"#\")\n"
     ]
    }
   ],
   "source": [
    "all_df = load_maf_data_2(file_path=constants.maf_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "01156b19-ae25-44c0-a11c-e807e16e99fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "      <th>Tumor_Validation_Allele2</th>\n",
       "      <th>Match_Norm_Validation_Allele1</th>\n",
       "      <th>Match_Norm_Validation_Allele2</th>\n",
       "      <th>Verification_Status</th>\n",
       "      <th>Validation_Status</th>\n",
       "      <th>Mutation_Status</th>\n",
       "      <th>Sequencing_Phase</th>\n",
       "      <th>Sequence_Source</th>\n",
       "      <th>Validation_Method</th>\n",
       "      <th>Score</th>\n",
       "      <th>BAM_File</th>\n",
       "      <th>Sequencer</th>\n",
       "      <th>Tumor_Sample_UUID</th>\n",
       "      <th>Matched_Norm_Sample_UUID</th>\n",
       "      <th>Genome_Change</th>\n",
       "      <th>Annotation_Transcript</th>\n",
       "      <th>Transcript_Strand</th>\n",
       "      <th>Transcript_Exon</th>\n",
       "      <th>Transcript_Position</th>\n",
       "      <th>cDNA_Change</th>\n",
       "      <th>Codon_Change</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>Other_Transcripts</th>\n",
       "      <th>Refseq_mRNA_Id</th>\n",
       "      <th>Refseq_prot_Id</th>\n",
       "      <th>SwissProt_acc_Id</th>\n",
       "      <th>SwissProt_entry_Id</th>\n",
       "      <th>Description</th>\n",
       "      <th>UniProt_AApos</th>\n",
       "      <th>UniProt_Region</th>\n",
       "      <th>UniProt_Site</th>\n",
       "      <th>UniProt_Natural_Variations</th>\n",
       "      <th>UniProt_Experimental_Info</th>\n",
       "      <th>GO_Biological_Process</th>\n",
       "      <th>GO_Cellular_Component</th>\n",
       "      <th>GO_Molecular_Function</th>\n",
       "      <th>COSMIC_overlapping_mutations</th>\n",
       "      <th>COSMIC_fusion_genes</th>\n",
       "      <th>COSMIC_tissue_types_affected</th>\n",
       "      <th>COSMIC_total_alterations_in_gene</th>\n",
       "      <th>Tumorscape_Amplification_Peaks</th>\n",
       "      <th>Tumorscape_Deletion_Peaks</th>\n",
       "      <th>TCGAscape_Amplification_Peaks</th>\n",
       "      <th>TCGAscape_Deletion_Peaks</th>\n",
       "      <th>DrugBank</th>\n",
       "      <th>ref_context</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>CCLE_ONCOMAP_overlapping_mutations</th>\n",
       "      <th>CCLE_ONCOMAP_total_mutations_in_gene</th>\n",
       "      <th>CGC_Mutation_Type</th>\n",
       "      <th>CGC_Translocation_Partner</th>\n",
       "      <th>CGC_Tumor_Types_Somatic</th>\n",
       "      <th>CGC_Tumor_Types_Germline</th>\n",
       "      <th>CGC_Other_Diseases</th>\n",
       "      <th>DNARepairGenes_Activity_linked_to_OMIM</th>\n",
       "      <th>FamilialCancerDatabase_Syndromes</th>\n",
       "      <th>MUTSIG_Published_Results</th>\n",
       "      <th>OREGANNO_ID</th>\n",
       "      <th>OREGANNO_Values</th>\n",
       "      <th>tumor_f</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>n_alt_count</th>\n",
       "      <th>n_ref_count</th>\n",
       "      <th>Gencode_34_secondaryVariantClassification</th>\n",
       "      <th>Achilles_Top_Genes</th>\n",
       "      <th>ClinVar_VCF_AF_ESP</th>\n",
       "      <th>ClinVar_VCF_AF_EXAC</th>\n",
       "      <th>ClinVar_VCF_AF_TGP</th>\n",
       "      <th>ClinVar_VCF_ALLELEID</th>\n",
       "      <th>ClinVar_VCF_CLNDISDB</th>\n",
       "      <th>ClinVar_VCF_CLNDISDBINCL</th>\n",
       "      <th>ClinVar_VCF_CLNDN</th>\n",
       "      <th>ClinVar_VCF_CLNDNINCL</th>\n",
       "      <th>ClinVar_VCF_CLNHGVS</th>\n",
       "      <th>ClinVar_VCF_CLNREVSTAT</th>\n",
       "      <th>ClinVar_VCF_CLNSIG</th>\n",
       "      <th>ClinVar_VCF_CLNSIGCONF</th>\n",
       "      <th>ClinVar_VCF_CLNSIGINCL</th>\n",
       "      <th>ClinVar_VCF_CLNVC</th>\n",
       "      <th>ClinVar_VCF_CLNVCSO</th>\n",
       "      <th>ClinVar_VCF_CLNVI</th>\n",
       "      <th>ClinVar_VCF_DBVARID</th>\n",
       "      <th>ClinVar_VCF_GENEINFO</th>\n",
       "      <th>ClinVar_VCF_MC</th>\n",
       "      <th>ClinVar_VCF_ORIGIN</th>\n",
       "      <th>ClinVar_VCF_RS</th>\n",
       "      <th>ClinVar_VCF_SSR</th>\n",
       "      <th>ClinVar_VCF_ID</th>\n",
       "      <th>ClinVar_VCF_FILTER</th>\n",
       "      <th>CosmicFusion_fusion_id</th>\n",
       "      <th>Familial_Cancer_Genes_Synonym</th>\n",
       "      <th>Familial_Cancer_Genes_Reference</th>\n",
       "      <th>Gencode_XHGNC_hgnc_id</th>\n",
       "      <th>HGNC_HGNC_ID</th>\n",
       "      <th>HGNC_Status</th>\n",
       "      <th>HGNC_Locus_Type</th>\n",
       "      <th>HGNC_Locus_Group</th>\n",
       "      <th>HGNC_Previous_Symbols</th>\n",
       "      <th>HGNC_Previous_Name</th>\n",
       "      <th>HGNC_Synonyms</th>\n",
       "      <th>HGNC_Name_Synonyms</th>\n",
       "      <th>HGNC_Chromosome</th>\n",
       "      <th>HGNC_Date_Modified</th>\n",
       "      <th>HGNC_Date_Symbol_Changed</th>\n",
       "      <th>HGNC_Date_Name_Changed</th>\n",
       "      <th>HGNC_Accession_Numbers</th>\n",
       "      <th>HGNC_Enzyme_IDs</th>\n",
       "      <th>HGNC_Ensembl_Gene_ID</th>\n",
       "      <th>HGNC_Pubmed_IDs</th>\n",
       "      <th>HGNC_RefSeq_IDs</th>\n",
       "      <th>HGNC_Gene_Family_ID</th>\n",
       "      <th>HGNC_Gene_Family_Name</th>\n",
       "      <th>HGNC_CCDS_IDs</th>\n",
       "      <th>HGNC_Vega_ID</th>\n",
       "      <th>HGNC_OMIM_ID(supplied_by_OMIM)</th>\n",
       "      <th>HGNC_RefSeq(supplied_by_NCBI)</th>\n",
       "      <th>HGNC_UniProt_ID(supplied_by_UniProt)</th>\n",
       "      <th>HGNC_Ensembl_ID(supplied_by_Ensembl)</th>\n",
       "      <th>HGNC_UCSC_ID(supplied_by_UCSC)</th>\n",
       "      <th>Oreganno_Build</th>\n",
       "      <th>Simple_Uniprot_alt_uniprot_accessions</th>\n",
       "      <th>dbSNP_ASP</th>\n",
       "      <th>dbSNP_ASS</th>\n",
       "      <th>dbSNP_CAF</th>\n",
       "      <th>dbSNP_CDA</th>\n",
       "      <th>dbSNP_CFL</th>\n",
       "      <th>dbSNP_COMMON</th>\n",
       "      <th>dbSNP_DSS</th>\n",
       "      <th>dbSNP_G5</th>\n",
       "      <th>dbSNP_G5A</th>\n",
       "      <th>dbSNP_GENEINFO</th>\n",
       "      <th>dbSNP_GNO</th>\n",
       "      <th>dbSNP_HD</th>\n",
       "      <th>dbSNP_INT</th>\n",
       "      <th>dbSNP_KGPhase1</th>\n",
       "      <th>dbSNP_KGPhase3</th>\n",
       "      <th>dbSNP_LSD</th>\n",
       "      <th>dbSNP_MTP</th>\n",
       "      <th>dbSNP_MUT</th>\n",
       "      <th>dbSNP_NOC</th>\n",
       "      <th>dbSNP_NOV</th>\n",
       "      <th>dbSNP_NSF</th>\n",
       "      <th>dbSNP_NSM</th>\n",
       "      <th>dbSNP_NSN</th>\n",
       "      <th>dbSNP_OM</th>\n",
       "      <th>dbSNP_OTH</th>\n",
       "      <th>dbSNP_PM</th>\n",
       "      <th>dbSNP_PMC</th>\n",
       "      <th>dbSNP_R3</th>\n",
       "      <th>dbSNP_R5</th>\n",
       "      <th>dbSNP_REF</th>\n",
       "      <th>dbSNP_RV</th>\n",
       "      <th>dbSNP_S3D</th>\n",
       "      <th>dbSNP_SAO</th>\n",
       "      <th>dbSNP_SLO</th>\n",
       "      <th>dbSNP_SSR</th>\n",
       "      <th>dbSNP_SYN</th>\n",
       "      <th>dbSNP_TOPMED</th>\n",
       "      <th>dbSNP_TPA</th>\n",
       "      <th>dbSNP_U3</th>\n",
       "      <th>dbSNP_U5</th>\n",
       "      <th>dbSNP_VC</th>\n",
       "      <th>dbSNP_VP</th>\n",
       "      <th>dbSNP_WGT</th>\n",
       "      <th>dbSNP_WTD</th>\n",
       "      <th>dbSNP_dbSNPBuildID</th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>dbSNP_FILTER</th>\n",
       "      <th>HGNC_Entrez_Gene_ID(supplied_by_NCBI)</th>\n",
       "      <th>dbSNP_RSPOS</th>\n",
       "      <th>dbSNP_VLD</th>\n",
       "      <th>AS_FilterStatus</th>\n",
       "      <th>AS_SB_TABLE</th>\n",
       "      <th>AS_UNIQ_ALT_READ_COUNT</th>\n",
       "      <th>CONTQ</th>\n",
       "      <th>DP</th>\n",
       "      <th>ECNT</th>\n",
       "      <th>GERMQ</th>\n",
       "      <th>MBQ</th>\n",
       "      <th>MFRL</th>\n",
       "      <th>MMQ</th>\n",
       "      <th>MPOS</th>\n",
       "      <th>NALOD</th>\n",
       "      <th>NCount</th>\n",
       "      <th>NLOD</th>\n",
       "      <th>OCM</th>\n",
       "      <th>PON</th>\n",
       "      <th>POPAF</th>\n",
       "      <th>ROQ</th>\n",
       "      <th>RPA</th>\n",
       "      <th>RU</th>\n",
       "      <th>SEQQ</th>\n",
       "      <th>STR</th>\n",
       "      <th>STRANDQ</th>\n",
       "      <th>STRQ</th>\n",
       "      <th>TLOD</th>\n",
       "      <th>Normal_Sample_Barcode</th>\n",
       "      <th>MutationID</th>\n",
       "      <th>QSS</th>\n",
       "      <th>TQSS</th>\n",
       "      <th>NT</th>\n",
       "      <th>QSS_NT</th>\n",
       "      <th>TQSS_NT</th>\n",
       "      <th>SGT</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>MQ</th>\n",
       "      <th>MQ0</th>\n",
       "      <th>ReadPosRankSum</th>\n",
       "      <th>SNVSB</th>\n",
       "      <th>PNOISE</th>\n",
       "      <th>PNOISE2</th>\n",
       "      <th>SomaticEVS</th>\n",
       "      <th>cosmic</th>\n",
       "      <th>ExAC_ALL</th>\n",
       "      <th>ExAC_AFR</th>\n",
       "      <th>ExAC_AMR</th>\n",
       "      <th>ExAC_EAS</th>\n",
       "      <th>ExAC_FIN</th>\n",
       "      <th>ExAC_NFE</th>\n",
       "      <th>ExAC_OTH</th>\n",
       "      <th>ExAC_SAS</th>\n",
       "      <th>AF</th>\n",
       "      <th>AF_raw</th>\n",
       "      <th>AF_male</th>\n",
       "      <th>AF_female</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_ami</th>\n",
       "      <th>AF_amr</th>\n",
       "      <th>AF_asj</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_oth</th>\n",
       "      <th>AF_sas</th>\n",
       "      <th>CLNALLELEID</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNREVSTAT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58350553</td>\n",
       "      <td>58350573</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CCAGGGCGAAGCGCGCGCCCT</td>\n",
       "      <td>CCAGGGCGAAGCGCGCGCCCT</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RESPOND_60100650</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>g.chr19:58350553_58350573delCCAGGGCGAAGCGCGCGCCCT</td>\n",
       "      <td>ENST00000263100.8</td>\n",
       "      <td>-</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1065_1085</td>\n",
       "      <td>c.989_1009delAGGGCGCGCGCTTCGCCCTGG</td>\n",
       "      <td>c.(988-1011)gagggcgcgcgcttcgccctggtg&gt;gtg</td>\n",
       "      <td>p.EGARFAL330del</td>\n",
       "      <td>AC012313.3_ENST00000599109.5_RNA|AC012313.3_EN...</td>\n",
       "      <td>NM_130786.3</td>\n",
       "      <td>NP_570602</td>\n",
       "      <td>P04217</td>\n",
       "      <td>A1BG_HUMAN</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood microparticle (GO:0072562)|extracellular...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biliary_tract(2)|breast(11)|central_nervous_sy...</td>\n",
       "      <td>91.0</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCCTCGCGCACCAGGGCGAAGCGCGCGCCCTCCAGGGGCGC</td>\n",
       "      <td>0.71734</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.028</td>\n",
       "      <td>4</td>\n",
       "      <td>179</td>\n",
       "      <td>0</td>\n",
       "      <td>189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EU794585</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>Approved</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>7/13/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>2591067</td>\n",
       "      <td>NM_130786</td>\n",
       "      <td>594</td>\n",
       "      <td>Immunoglobulin like domain containing</td>\n",
       "      <td>CCDS12976</td>\n",
       "      <td>OTTHUMG00000183507</td>\n",
       "      <td>138670</td>\n",
       "      <td>NM_130786</td>\n",
       "      <td>P04217</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>uc002qsd.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A8K052|Q68CK0|Q8IYJ6|Q96P39</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>SITE</td>\n",
       "      <td>[131, 237|2, 2]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>453.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>[20, 20]</td>\n",
       "      <td>[193, 176]</td>\n",
       "      <td>[60, 60]</td>\n",
       "      <td>58.0</td>\n",
       "      <td>2.21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.2</td>\n",
       "      <td>RESPOND_60100650_C</td>\n",
       "      <td>RESPOND_60100650_chr19_A1BG_58350553</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>exonic</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id       Center NCBI_Build Chromosome  \\\n",
       "0        A1BG             1.0  __UNKNOWN__       hg38      chr19   \n",
       "\n",
       "   Start_Position  End_Position Strand Variant_Classification Variant_Type  \\\n",
       "0        58350553      58350573      +           In_Frame_Del          DEL   \n",
       "\n",
       "        Reference_Allele      Tumor_Seq_Allele1 Tumor_Seq_Allele2 dbSNP_RS  \\\n",
       "0  CCAGGGCGAAGCGCGCGCCCT  CCAGGGCGAAGCGCGCGCCCT                 -      NaN   \n",
       "\n",
       "  dbSNP_Val_Status Tumor_Sample_Barcode Matched_Norm_Sample_Barcode  \\\n",
       "0              NaN     RESPOND_60100650                 __UNKNOWN__   \n",
       "\n",
       "  Match_Norm_Seq_Allele1 Match_Norm_Seq_Allele2 Tumor_Validation_Allele1  \\\n",
       "0                      -                      -              __UNKNOWN__   \n",
       "\n",
       "  Tumor_Validation_Allele2 Match_Norm_Validation_Allele1  \\\n",
       "0              __UNKNOWN__                   __UNKNOWN__   \n",
       "\n",
       "  Match_Norm_Validation_Allele2 Verification_Status Validation_Status  \\\n",
       "0                   __UNKNOWN__         __UNKNOWN__       __UNKNOWN__   \n",
       "\n",
       "  Mutation_Status Sequencing_Phase Sequence_Source Validation_Method  Score  \\\n",
       "0     __UNKNOWN__      __UNKNOWN__     __UNKNOWN__       __UNKNOWN__    NaN   \n",
       "\n",
       "   BAM_File    Sequencer Tumor_Sample_UUID Matched_Norm_Sample_UUID  \\\n",
       "0       NaN  __UNKNOWN__       __UNKNOWN__              __UNKNOWN__   \n",
       "\n",
       "                                       Genome_Change Annotation_Transcript  \\\n",
       "0  g.chr19:58350553_58350573delCCAGGGCGAAGCGCGCGCCCT     ENST00000263100.8   \n",
       "\n",
       "  Transcript_Strand  Transcript_Exon Transcript_Position  \\\n",
       "0                 -              6.0           1065_1085   \n",
       "\n",
       "                          cDNA_Change  \\\n",
       "0  c.989_1009delAGGGCGCGCGCTTCGCCCTGG   \n",
       "\n",
       "                               Codon_Change   Protein_Change  \\\n",
       "0  c.(988-1011)gagggcgcgcgcttcgccctggtg>gtg  p.EGARFAL330del   \n",
       "\n",
       "                                   Other_Transcripts Refseq_mRNA_Id  \\\n",
       "0  AC012313.3_ENST00000599109.5_RNA|AC012313.3_EN...    NM_130786.3   \n",
       "\n",
       "  Refseq_prot_Id SwissProt_acc_Id SwissProt_entry_Id             Description  \\\n",
       "0      NP_570602           P04217         A1BG_HUMAN  alpha-1-B glycoprotein   \n",
       "\n",
       "  UniProt_AApos UniProt_Region UniProt_Site UniProt_Natural_Variations  \\\n",
       "0   __UNKNOWN__    __UNKNOWN__  __UNKNOWN__                __UNKNOWN__   \n",
       "\n",
       "  UniProt_Experimental_Info GO_Biological_Process  \\\n",
       "0               __UNKNOWN__                   NaN   \n",
       "\n",
       "                               GO_Cellular_Component GO_Molecular_Function  \\\n",
       "0  blood microparticle (GO:0072562)|extracellular...                   NaN   \n",
       "\n",
       "  COSMIC_overlapping_mutations COSMIC_fusion_genes  \\\n",
       "0                          NaN                 NaN   \n",
       "\n",
       "                        COSMIC_tissue_types_affected  \\\n",
       "0  biliary_tract(2)|breast(11)|central_nervous_sy...   \n",
       "\n",
       "   COSMIC_total_alterations_in_gene Tumorscape_Amplification_Peaks  \\\n",
       "0                              91.0                    __UNKNOWN__   \n",
       "\n",
       "  Tumorscape_Deletion_Peaks TCGAscape_Amplification_Peaks  \\\n",
       "0               __UNKNOWN__                   __UNKNOWN__   \n",
       "\n",
       "  TCGAscape_Deletion_Peaks DrugBank  \\\n",
       "0              __UNKNOWN__      NaN   \n",
       "\n",
       "                                 ref_context  gc_content  \\\n",
       "0  TCCTCGCGCACCAGGGCGAAGCGCGCGCCCTCCAGGGGCGC     0.71734   \n",
       "\n",
       "  CCLE_ONCOMAP_overlapping_mutations CCLE_ONCOMAP_total_mutations_in_gene  \\\n",
       "0                        __UNKNOWN__                          __UNKNOWN__   \n",
       "\n",
       "  CGC_Mutation_Type CGC_Translocation_Partner CGC_Tumor_Types_Somatic  \\\n",
       "0       __UNKNOWN__               __UNKNOWN__             __UNKNOWN__   \n",
       "\n",
       "  CGC_Tumor_Types_Germline CGC_Other_Diseases  \\\n",
       "0              __UNKNOWN__        __UNKNOWN__   \n",
       "\n",
       "  DNARepairGenes_Activity_linked_to_OMIM FamilialCancerDatabase_Syndromes  \\\n",
       "0                            __UNKNOWN__                              NaN   \n",
       "\n",
       "  MUTSIG_Published_Results OREGANNO_ID OREGANNO_Values  tumor_f  t_alt_count  \\\n",
       "0              __UNKNOWN__         NaN             NaN    0.028            4   \n",
       "\n",
       "   t_ref_count  n_alt_count  n_ref_count  \\\n",
       "0          179            0          189   \n",
       "\n",
       "  Gencode_34_secondaryVariantClassification Achilles_Top_Genes  \\\n",
       "0                                       NaN                NaN   \n",
       "\n",
       "   ClinVar_VCF_AF_ESP  ClinVar_VCF_AF_EXAC  ClinVar_VCF_AF_TGP  \\\n",
       "0                 NaN                  NaN                 NaN   \n",
       "\n",
       "   ClinVar_VCF_ALLELEID ClinVar_VCF_CLNDISDB ClinVar_VCF_CLNDISDBINCL  \\\n",
       "0                   NaN                  NaN                      NaN   \n",
       "\n",
       "  ClinVar_VCF_CLNDN ClinVar_VCF_CLNDNINCL ClinVar_VCF_CLNHGVS  \\\n",
       "0               NaN                   NaN                 NaN   \n",
       "\n",
       "  ClinVar_VCF_CLNREVSTAT ClinVar_VCF_CLNSIG ClinVar_VCF_CLNSIGCONF  \\\n",
       "0                    NaN                NaN                    NaN   \n",
       "\n",
       "  ClinVar_VCF_CLNSIGINCL ClinVar_VCF_CLNVC ClinVar_VCF_CLNVCSO  \\\n",
       "0                    NaN               NaN                 NaN   \n",
       "\n",
       "  ClinVar_VCF_CLNVI  ClinVar_VCF_DBVARID ClinVar_VCF_GENEINFO ClinVar_VCF_MC  \\\n",
       "0               NaN                  NaN                  NaN            NaN   \n",
       "\n",
       "   ClinVar_VCF_ORIGIN  ClinVar_VCF_RS  ClinVar_VCF_SSR  ClinVar_VCF_ID  \\\n",
       "0                 NaN             NaN              NaN             NaN   \n",
       "\n",
       "   ClinVar_VCF_FILTER  CosmicFusion_fusion_id Familial_Cancer_Genes_Synonym  \\\n",
       "0                 NaN                     NaN                           NaN   \n",
       "\n",
       "  Familial_Cancer_Genes_Reference Gencode_XHGNC_hgnc_id HGNC_HGNC_ID  \\\n",
       "0                             NaN              EU794585       HGNC:5   \n",
       "\n",
       "  HGNC_Status            HGNC_Locus_Type     HGNC_Locus_Group  \\\n",
       "0    Approved  gene with protein product  protein-coding gene   \n",
       "\n",
       "  HGNC_Previous_Symbols HGNC_Previous_Name HGNC_Synonyms HGNC_Name_Synonyms  \\\n",
       "0                   NaN                NaN           NaN                NaN   \n",
       "\n",
       "  HGNC_Chromosome HGNC_Date_Modified HGNC_Date_Symbol_Changed  \\\n",
       "0        19q13.43            7/13/15                      NaN   \n",
       "\n",
       "  HGNC_Date_Name_Changed HGNC_Accession_Numbers HGNC_Enzyme_IDs  \\\n",
       "0                    NaN                    NaN             NaN   \n",
       "\n",
       "  HGNC_Ensembl_Gene_ID HGNC_Pubmed_IDs HGNC_RefSeq_IDs HGNC_Gene_Family_ID  \\\n",
       "0      ENSG00000121410         2591067       NM_130786                 594   \n",
       "\n",
       "                   HGNC_Gene_Family_Name HGNC_CCDS_IDs        HGNC_Vega_ID  \\\n",
       "0  Immunoglobulin like domain containing     CCDS12976  OTTHUMG00000183507   \n",
       "\n",
       "  HGNC_OMIM_ID(supplied_by_OMIM) HGNC_RefSeq(supplied_by_NCBI)  \\\n",
       "0                         138670                     NM_130786   \n",
       "\n",
       "  HGNC_UniProt_ID(supplied_by_UniProt) HGNC_Ensembl_ID(supplied_by_Ensembl)  \\\n",
       "0                               P04217                      ENSG00000121410   \n",
       "\n",
       "  HGNC_UCSC_ID(supplied_by_UCSC) Oreganno_Build  \\\n",
       "0                     uc002qsd.5            NaN   \n",
       "\n",
       "  Simple_Uniprot_alt_uniprot_accessions dbSNP_ASP dbSNP_ASS dbSNP_CAF  \\\n",
       "0           A8K052|Q68CK0|Q8IYJ6|Q96P39     False     False       NaN   \n",
       "\n",
       "  dbSNP_CDA dbSNP_CFL dbSNP_COMMON dbSNP_DSS dbSNP_G5 dbSNP_G5A  \\\n",
       "0     False     False          NaN     False    False     False   \n",
       "\n",
       "  dbSNP_GENEINFO dbSNP_GNO dbSNP_HD dbSNP_INT dbSNP_KGPhase1 dbSNP_KGPhase3  \\\n",
       "0            NaN     False    False     False          False          False   \n",
       "\n",
       "  dbSNP_LSD dbSNP_MTP dbSNP_MUT dbSNP_NOC dbSNP_NOV dbSNP_NSF dbSNP_NSM  \\\n",
       "0     False     False     False     False     False     False     False   \n",
       "\n",
       "  dbSNP_NSN dbSNP_OM dbSNP_OTH dbSNP_PM dbSNP_PMC dbSNP_R3 dbSNP_R5 dbSNP_REF  \\\n",
       "0     False    False     False    False     False    False    False     False   \n",
       "\n",
       "  dbSNP_RV dbSNP_S3D dbSNP_SAO dbSNP_SLO dbSNP_SSR dbSNP_SYN dbSNP_TOPMED  \\\n",
       "0    False     False       NaN     False       NaN     False          NaN   \n",
       "\n",
       "  dbSNP_TPA dbSNP_U3 dbSNP_U5 dbSNP_VC dbSNP_VP dbSNP_WGT dbSNP_WTD  \\\n",
       "0     False    False    False      NaN      NaN       NaN     False   \n",
       "\n",
       "  dbSNP_dbSNPBuildID dbSNP_ID dbSNP_FILTER  \\\n",
       "0                NaN      NaN          NaN   \n",
       "\n",
       "   HGNC_Entrez_Gene_ID(supplied_by_NCBI) dbSNP_RSPOS dbSNP_VLD  \\\n",
       "0                                    1.0         NaN     False   \n",
       "\n",
       "  AS_FilterStatus      AS_SB_TABLE  AS_UNIQ_ALT_READ_COUNT  CONTQ     DP  \\\n",
       "0            SITE  [131, 237|2, 2]                     NaN    NaN  453.0   \n",
       "\n",
       "   ECNT  GERMQ       MBQ        MFRL       MMQ  MPOS  NALOD  NCount   NLOD  \\\n",
       "0   2.0   93.0  [20, 20]  [193, 176]  [60, 60]  58.0   2.21     NaN  42.06   \n",
       "\n",
       "   OCM  PON  POPAF   ROQ  RPA   RU  SEQQ  STR  STRANDQ  STRQ  TLOD  \\\n",
       "0  NaN  NaN    6.0  93.0  NaN  NaN   NaN  NaN      NaN   NaN  12.2   \n",
       "\n",
       "  Normal_Sample_Barcode                            MutationID  QSS  TQSS   NT  \\\n",
       "0    RESPOND_60100650_C  RESPOND_60100650_chr19_A1BG_58350553  NaN   NaN  NaN   \n",
       "\n",
       "   QSS_NT  TQSS_NT  SGT SOMATIC  MQ  MQ0  ReadPosRankSum  SNVSB  PNOISE  \\\n",
       "0     NaN      NaN  NaN     NaN NaN  NaN             NaN    NaN     NaN   \n",
       "\n",
       "   PNOISE2  SomaticEVS cosmic ExAC_ALL ExAC_AFR ExAC_AMR ExAC_EAS ExAC_FIN  \\\n",
       "0      NaN         NaN      .        .        .        .        .        .   \n",
       "\n",
       "  ExAC_NFE ExAC_OTH ExAC_SAS AF AF_raw AF_male AF_female AF_afr AF_ami AF_amr  \\\n",
       "0        .        .        .  .      .       .         .      .      .      .   \n",
       "\n",
       "  AF_asj AF_eas AF_fin AF_nfe AF_oth AF_sas CLNALLELEID CLNDN CLNDISDB  \\\n",
       "0      .      .      .      .      .      .           .     .        .   \n",
       "\n",
       "  CLNREVSTAT CLNSIG Func.refGene Gene.refGene GeneDetail.refGene  \\\n",
       "0          .      .       exonic         A1BG                  .   \n",
       "\n",
       "  ExonicFunc.refGene AAChange.refGene  mutation  \n",
       "0                  .                .         1  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_df.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfc3e66a-555e-4ff9-acd7-0dc92ce01647",
   "metadata": {},
   "source": [
    "## "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43f94922-9d3a-4049-8d75-ea01d1d75697",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
