{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a68179f1-6218-4af0-b352-79dccdb72e82",
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_file_path = 'data/RESPOND_247_coding_final.maf'\n",
    "expression_file_path = 'data/Expression_remove_BE.txt'\n",
    "target_gene='SPOP'\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statistics import mean \n",
    "from scipy.stats import ranksums\n",
    "import matplotlib.pylab as plt\n",
    "from statsmodels.stats import multitest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "12bdca97-76e8-4627-8c49-51098cc5fd8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_maf_data(file_path):\n",
    "    # Define columns of interest\n",
    "    columns = [\"Hugo_Symbol\", \"Tumor_Sample_Barcode\"]\n",
    "    df = pd.read_csv(file_path, sep='\\t', comment=\"#\", usecols=columns)\n",
    "    df.rename(columns={'Hugo_Symbol': 'gene', 'Tumor_Sample_Barcode': 'sample'}, inplace=True)\n",
    "    df['mutation'] = 1\n",
    "    return df\n",
    "    \n",
    "maf_df = load_maf_data(maf_file_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "3561f69f-7e5d-4c57-b6a8-befedcba5104",
   "metadata": {},
   "outputs": [],
   "source": [
    "def log2(value):\n",
    "    return np.log2(value + 1)\n",
    "\n",
    "def load_txt_file_into_dataframe(file_path):\n",
    "    # Read the .txt file into a pandas DataFrame\n",
    "    df = pd.read_csv(file_path, sep='\\t')  # Adjust the separator if needed\n",
    "\n",
    "    # take log2 of expression data to scale expression data\n",
    "    df = df.map(log2)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "# Call the load_txt_file_into_dataframe function\n",
    "expression_df = load_txt_file_into_dataframe(expression_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "c957ea88-e994-4193-9e16-35cb3c73ba59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reformat_expression_data(df):\n",
    "    # Combine column names and index names into rows for every element\n",
    "    melted_df = pd.melt(df.reset_index(), id_vars=['index'], var_name='column', col_level=0)\n",
    "\n",
    "    # Rename columns\n",
    "    melted_df.rename(columns={'index': 'gene', 'column': 'sample', 'value': 'gene_expression'}, inplace=True)\n",
    "\n",
    "    return melted_df\n",
    "\n",
    "\n",
    "expression_df_melted = reformat_expression_data(expression_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "6d348536-277e-465d-b461-25a9035ff9ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# note individuals with muttated data may be missing in the expression data and vice versa. \n",
    "# every sample should have at least one mutation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "3372f141-9ba1-4e9f-9aaf-adcf16eb43c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fraction of rows filtered is 0.2865853658536585\n"
     ]
    }
   ],
   "source": [
    "def preprocess_and_combine_mutation_expression(maf_df, expression_df):\n",
    "    ''' filter the expression data to those that have whole genome sequencing \n",
    "     i.e. appear in the mutation data frame (maf)\n",
    "     join mutation and expression data\n",
    "     '''\n",
    "    \n",
    "    exon_seq_samples = maf_df['sample'].unique()\n",
    "    filtered_expression_df = expression_df[expression_df['sample'].isin(exon_seq_samples)]\n",
    "    \n",
    "    all_rows = expression_df.shape[0]\n",
    "    filt_rows = filtered_expression_df.shape[0]\n",
    "    \n",
    "    percentage_filtered = (all_rows - filt_rows) / all_rows\n",
    "    \n",
    "    print('fraction of rows filtered is', percentage_filtered)\n",
    "\n",
    "    express_mut_genes_df = pd.merge(maf_df, filtered_expression_df, on=['gene', 'sample'], how='right')\n",
    "\n",
    "    express_mut_genes_df['mutation'].fillna(0, inplace=True)\n",
    "\n",
    "    return express_mut_genes_df\n",
    "\n",
    "\n",
    "mutation_expression_df_melted =  preprocess_and_combine_mutation_expression(maf_df, expression_df_melted)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "3a6df317-d9bc-492d-8c4f-03186375ada3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_log_fold(list_mutated, list_non_mutated, smoothing_factor=1e-8): \n",
    "    mean_mutated = mean(list_mutated)  + smoothing_factor\n",
    "    mean_non_mutated = mean(list_non_mutated) + smoothing_factor\n",
    "        \n",
    "    return np.log2(mean_mutated/mean_non_mutated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "377f6fb9-7597-411b-a810-af33c7f829eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_z_score(list_mutated, list_non_mutated):\n",
    "    mean_mutated = mean(list_mutated) \n",
    "    mean_non_mutated = mean(list_non_mutated) \n",
    "    all_expression_values = list_non_mutated + list_mutated\n",
    "    z_score = (mean_mutated - mean_non_mutated) / np.std(all_expression_values)\n",
    "    return z_score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "36736785-ca3e-4a5e-8b34-d550cef0d3ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this isn't used for gene expression \n",
    "# from scipy.stats import ttest_ind\n",
    "# def calculate_pvalue_and_effect_size_ttest(list_mutated, list_non_mutated):\n",
    "#     t_stat, p_value = ttest_ind(list_mutated, list_non_mutated, equal_var=False)\n",
    "#     return p_value, t_stat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "7aa8e148-4bb5-4bff-9cb3-7ae71e4bb2ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_pvalue_and_effect_size_wilcox_ranksum(list_mutated, list_non_mutated):\n",
    "     test = ranksums(list_mutated, list_non_mutated)\n",
    "     u_statistic = test.statistic \n",
    "     effect_size = u_statistic / (len(list_mutated) * len(list_non_mutated))\n",
    "     pvalue=test.pvalue\n",
    "     return pvalue, effect_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "0851fd78-7875-49c2-ace0-98a8be179d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_adjusted_pvalue(pvalues, method='fdr_bh'):\n",
    "    _, corrected_pvalues, _, _ = multitest.multipletests(pvalues, method=method)\n",
    "    return corrected_pvalues\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "78f10ce4-a336-4741-944c-7642a31e304f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SPOP'"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "231c6d1e-4a02-45bf-943c-b8ddb9c76757",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "def generate_stats_per_gene(express_mut_genes_df, target_gene, output_folder=target_gene):\n",
    "    \"\"\"\n",
    "    Given expression and mutation data calculates LogFC, pvalue, mean of expression per gene \n",
    "    for a given target_gene \n",
    "    \n",
    "    \"\"\"\n",
    "    gene_df = express_mut_genes_df[express_mut_genes_df['gene']==target_gene]\n",
    "\n",
    "    if len(gene_df) == 0:\n",
    "        raise ValueError(\"This gene is not valid! No mutations with this gene exist\") \n",
    "\n",
    "    mutated_samples = gene_df[gene_df['mutation'] == 1]['sample']\n",
    "    non_mutated_samples = gene_df[gene_df['mutation'] == 0]['sample']\n",
    "    \n",
    "    mutated_individuals_expression = express_mut_genes_df[express_mut_genes_df['sample'].isin(mutated_samples)]\n",
    "    \n",
    "    non_mutated_individuals_expression = express_mut_genes_df[express_mut_genes_df['sample'].isin(non_mutated_samples)]\n",
    "\n",
    "    # gather data of mutated and non-mutated genes into lists\n",
    "    mutated_individuals_data = mutated_individuals_expression.groupby(['gene'])['gene_expression'].apply(\n",
    "    lambda x: list(x)).to_frame().reset_index().rename(columns={'gene_expression': 'gene_expression_mutated'})\n",
    "    non_mutated_individuals_data = non_mutated_individuals_expression.groupby(['gene'])['gene_expression'].apply(\n",
    "    lambda x: list(x)).to_frame().reset_index().rename(columns={'gene_expression': 'gene_expression_non_mutated'})\n",
    "\n",
    "    # combine mutated and unmutated data into one df\n",
    "    combined_data= pd.merge(mutated_individuals_data, non_mutated_individuals_data, on='gene', how='inner')\n",
    "\n",
    "    # calculate fold change and p-value and z-score\n",
    "    combined_data['logFC']= combined_data.apply(\n",
    "        lambda x: calculate_log_fold(x.gene_expression_mutated, x.gene_expression_non_mutated), axis=1)\n",
    "    # calculate wilcox pvalue\n",
    "    combined_data[['pvalue', 'effect_size']] = combined_data.apply(\n",
    "    lambda x: pd.Series(calculate_pvalue_and_effect_size_wilcox_ranksum(x.gene_expression_mutated, x.gene_expression_non_mutated)),\n",
    "    axis=1\n",
    "    )\n",
    "    combined_data['expression_mutated_mean'] = combined_data['gene_expression_mutated'].apply(mean)\n",
    "    combined_data['expression_nonmutated_mean'] = combined_data['gene_expression_non_mutated'].apply(mean)\n",
    "\n",
    "\n",
    "    # calculate adjusted p-value\n",
    "    combined_data['adjusted_pvalue'] = calculate_adjusted_pvalue(combined_data['pvalue'].values)\n",
    "\n",
    "    # Output data to csv \n",
    "    combined_data.drop(columns=['gene_expression_mutated', 'gene_expression_non_mutated'], inplace=True)\n",
    "    \n",
    "    os.makedirs(output_folder, exist_ok=True)\n",
    "    output_filename = f'{output_folder}/{mutated_samples.count()}_{non_mutated_samples.count()}_logfc_pvalue.csv'\n",
    "    \n",
    "    print(f\"outputting data to {output_filename}\")\n",
    "    combined_data.to_csv(output_filename, index=False)\n",
    "\n",
    "    return combined_data, mutated_samples, output_filename\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "ab65d785-dbfe-4336-a4d0-68024188234c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputting data to SPOP/15_102_logfc_pvalue.csv\n"
     ]
    }
   ],
   "source": [
    "volcano_plot_df, individuals_mutated_target_gene, volcano_input_filename = generate_stats_per_gene(\n",
    "    mutation_expression_df_melted, \n",
    "    target_gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e484ffc8-6249-448f-b354-01574bb19717",
   "metadata": {},
   "source": [
    "## generate heatmap of top 100 differentially expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "fe591289-eba9-427a-8563-ada51b200b91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_expression_heatmap(expression_df, volcano_plot_df, n=100, exclude_value=10):\n",
    "    # Get the indices of the top n rows based on absolute values of 'logFC'\n",
    "    # get the genes most differentially expressed (high log FC values)\n",
    "    # if we exclude -exclude_value, exclude_value values\n",
    "    # top_n_indices = volcano_plot_df[\n",
    "    #     (volcano_plot_df['logFC'].abs() != exclude_value)\n",
    "    # ]['logFC'].abs().nlargest(n).index\n",
    "    \n",
    "    # top_n_rows = volcano_plot_df.loc[top_n_indices].set_index('gene')\n",
    "\n",
    "    top_n_indices = volcano_plot_df['pvalue'].nsmallest(n).index\n",
    "    \n",
    "    top_n_rows = volcano_plot_df.loc[top_n_indices].set_index('gene')\n",
    "\n",
    "    # go back to the expression df and make a heatmap of the top n genes\n",
    "    expression_df_heatmap = expression_df.loc[top_n_rows.index]\n",
    "    \n",
    "    return expression_df_heatmap\n",
    "\n",
    "heatmap_data = generate_expression_heatmap(expression_df, volcano_plot_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "e51e4364-9eec-4ef8-ba66-16410c606ac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SPOP/sample_mutation_status.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_mutated_status(expression_df_heatmap, individuals_mutated_target_gene, output_folder=target_gene):\n",
    "    individuals_mutated_target_gene = list(individuals_mutated_target_gene)\n",
    "    \n",
    "    mutated_status = [1 if item in individuals_mutated_target_gene else 0 for item in expression_df_heatmap.columns]\n",
    "\n",
    "    sample_categories_df = pd.DataFrame({\n",
    "        'Sample': expression_df_heatmap.columns,\n",
    "        'Mutation Status': mutated_status\n",
    "    })\n",
    "\n",
    "    output_filename = f'{output_folder}/sample_mutation_status.csv'\n",
    "    print(output_filename)\n",
    "    sample_categories_df.to_csv(output_filename,  index=False)\n",
    "    return sample_categories_df\n",
    "\n",
    "mutated_status_df = get_mutated_status(heatmap_data, individuals_mutated_target_gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "9af3d157-2ce3-4456-a6d9-a7ecbbaab5c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RESPOND_10100218</th>\n",
       "      <th>RESPOND_10100291</th>\n",
       "      <th>RESPOND_10100412</th>\n",
       "      <th>RESPOND_10100478</th>\n",
       "      <th>RESPOND_10100596</th>\n",
       "      <th>RESPOND_10100615</th>\n",
       "      <th>RESPOND_10100801</th>\n",
       "      <th>RESPOND_10100884</th>\n",
       "      <th>RESPOND_10100899</th>\n",
       "      <th>RESPOND_10100952</th>\n",
       "      <th>...</th>\n",
       "      <th>RESPOND_80100242</th>\n",
       "      <th>RESPOND_80100313</th>\n",
       "      <th>RESPOND_80100345</th>\n",
       "      <th>RESPOND_80100411</th>\n",
       "      <th>RESPOND_80100526</th>\n",
       "      <th>RESPOND_80100556</th>\n",
       "      <th>RESPOND_80100590</th>\n",
       "      <th>RESPOND_81100031</th>\n",
       "      <th>RESPOND_40100842</th>\n",
       "      <th>RESPOND_80100259</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARHGEF16</th>\n",
       "      <td>5.554589</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>5.754888</td>\n",
       "      <td>4.321928</td>\n",
       "      <td>6.672425</td>\n",
       "      <td>4.087463</td>\n",
       "      <td>5.977280</td>\n",
       "      <td>6.614710</td>\n",
       "      <td>4.807355</td>\n",
       "      <td>4.700440</td>\n",
       "      <td>...</td>\n",
       "      <td>6.768184</td>\n",
       "      <td>5.754888</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>7.179909</td>\n",
       "      <td>6.794416</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>5.169925</td>\n",
       "      <td>6.285402</td>\n",
       "      <td>5.807355</td>\n",
       "      <td>6.303781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMG20B</th>\n",
       "      <td>10.707359</td>\n",
       "      <td>8.781360</td>\n",
       "      <td>9.667112</td>\n",
       "      <td>8.426265</td>\n",
       "      <td>9.915879</td>\n",
       "      <td>9.308339</td>\n",
       "      <td>10.314017</td>\n",
       "      <td>8.044394</td>\n",
       "      <td>8.361944</td>\n",
       "      <td>8.299208</td>\n",
       "      <td>...</td>\n",
       "      <td>10.392317</td>\n",
       "      <td>10.890264</td>\n",
       "      <td>8.499846</td>\n",
       "      <td>11.106563</td>\n",
       "      <td>9.874981</td>\n",
       "      <td>9.154818</td>\n",
       "      <td>9.491853</td>\n",
       "      <td>9.292322</td>\n",
       "      <td>9.296916</td>\n",
       "      <td>10.213104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLXNB2</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>9.103288</td>\n",
       "      <td>9.623881</td>\n",
       "      <td>9.388017</td>\n",
       "      <td>9.977280</td>\n",
       "      <td>8.918863</td>\n",
       "      <td>9.262095</td>\n",
       "      <td>9.074141</td>\n",
       "      <td>8.686501</td>\n",
       "      <td>9.169925</td>\n",
       "      <td>...</td>\n",
       "      <td>9.611025</td>\n",
       "      <td>10.292322</td>\n",
       "      <td>8.810572</td>\n",
       "      <td>10.650154</td>\n",
       "      <td>10.079485</td>\n",
       "      <td>8.625709</td>\n",
       "      <td>9.337622</td>\n",
       "      <td>9.612868</td>\n",
       "      <td>9.317413</td>\n",
       "      <td>9.569856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WFIKKN1</th>\n",
       "      <td>2.321928</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.857649</td>\n",
       "      <td>...</td>\n",
       "      <td>3.906891</td>\n",
       "      <td>3.321928</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.459432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MC1R</th>\n",
       "      <td>4.459432</td>\n",
       "      <td>7.330917</td>\n",
       "      <td>6.228819</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>6.266787</td>\n",
       "      <td>4.523562</td>\n",
       "      <td>4.906891</td>\n",
       "      <td>6.228819</td>\n",
       "      <td>5.754888</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>...</td>\n",
       "      <td>6.629357</td>\n",
       "      <td>6.523562</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>7.965784</td>\n",
       "      <td>7.483816</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>7.523562</td>\n",
       "      <td>6.392317</td>\n",
       "      <td>6.209453</td>\n",
       "      <td>7.475733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBD3</th>\n",
       "      <td>6.599913</td>\n",
       "      <td>6.189825</td>\n",
       "      <td>7.339850</td>\n",
       "      <td>6.108524</td>\n",
       "      <td>8.022368</td>\n",
       "      <td>6.321928</td>\n",
       "      <td>6.870365</td>\n",
       "      <td>7.651052</td>\n",
       "      <td>5.357552</td>\n",
       "      <td>6.247928</td>\n",
       "      <td>...</td>\n",
       "      <td>7.033423</td>\n",
       "      <td>7.434628</td>\n",
       "      <td>5.169925</td>\n",
       "      <td>8.370687</td>\n",
       "      <td>8.118941</td>\n",
       "      <td>4.954196</td>\n",
       "      <td>7.011227</td>\n",
       "      <td>7.977280</td>\n",
       "      <td>6.700440</td>\n",
       "      <td>7.033423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYB561D1</th>\n",
       "      <td>3.459432</td>\n",
       "      <td>5.087463</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>4.087463</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>2.807355</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>4.321928</td>\n",
       "      <td>4.087463</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.906891</td>\n",
       "      <td>3.906891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GUCY1B1</th>\n",
       "      <td>3.169925</td>\n",
       "      <td>3.321928</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>2.807355</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.906891</td>\n",
       "      <td>4.247928</td>\n",
       "      <td>4.169925</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>4.392317</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>2.807355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF1</th>\n",
       "      <td>2.321928</td>\n",
       "      <td>0.371710</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.931392</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044200</td>\n",
       "      <td>0.449118</td>\n",
       "      <td>0.203308</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.102749</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.852043</td>\n",
       "      <td>0.879822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042794</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.078879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEGF6</th>\n",
       "      <td>3.584963</td>\n",
       "      <td>6.266787</td>\n",
       "      <td>5.672425</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>6.375039</td>\n",
       "      <td>4.754888</td>\n",
       "      <td>4.906891</td>\n",
       "      <td>6.087463</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>5.209453</td>\n",
       "      <td>...</td>\n",
       "      <td>6.894818</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>6.087463</td>\n",
       "      <td>5.247928</td>\n",
       "      <td>5.129283</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.087463</td>\n",
       "      <td>7.022368</td>\n",
       "      <td>6.022368</td>\n",
       "      <td>5.285402</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 164 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          RESPOND_10100218  RESPOND_10100291  RESPOND_10100412  \\\n",
       "gene                                                             \n",
       "ARHGEF16          5.554589          5.426265          5.754888   \n",
       "HMG20B           10.707359          8.781360          9.667112   \n",
       "PLXNB2           10.000000          9.103288          9.623881   \n",
       "WFIKKN1           2.321928          3.459432          3.169925   \n",
       "MC1R              4.459432          7.330917          6.228819   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              6.599913          6.189825          7.339850   \n",
       "CYB561D1          3.459432          5.087463          4.000000   \n",
       "GUCY1B1           3.169925          3.321928          3.000000   \n",
       "KLF1              2.321928          0.371710          0.000000   \n",
       "MEGF6             3.584963          6.266787          5.672425   \n",
       "\n",
       "          RESPOND_10100478  RESPOND_10100596  RESPOND_10100615  \\\n",
       "gene                                                             \n",
       "ARHGEF16          4.321928          6.672425          4.087463   \n",
       "HMG20B            8.426265          9.915879          9.308339   \n",
       "PLXNB2            9.388017          9.977280          8.918863   \n",
       "WFIKKN1           2.321928          3.584963          1.584963   \n",
       "MC1R              5.000000          6.266787          4.523562   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              6.108524          8.022368          6.321928   \n",
       "CYB561D1          3.459432          4.087463          3.700440   \n",
       "GUCY1B1           5.426265          2.807355          3.459432   \n",
       "KLF1              0.000000          0.931392          0.000000   \n",
       "MEGF6             4.584963          6.375039          4.754888   \n",
       "\n",
       "          RESPOND_10100801  RESPOND_10100884  RESPOND_10100899  \\\n",
       "gene                                                             \n",
       "ARHGEF16          5.977280          6.614710          4.807355   \n",
       "HMG20B           10.314017          8.044394          8.361944   \n",
       "PLXNB2            9.262095          9.074141          8.686501   \n",
       "WFIKKN1           3.000000          3.000000          2.000000   \n",
       "MC1R              4.906891          6.228819          5.754888   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              6.870365          7.651052          5.357552   \n",
       "CYB561D1          3.459432          3.807355          3.169925   \n",
       "GUCY1B1           3.906891          4.247928          4.169925   \n",
       "KLF1              0.044200          0.449118          0.203308   \n",
       "MEGF6             4.906891          6.087463          4.584963   \n",
       "\n",
       "          RESPOND_10100952  ...  RESPOND_80100242  RESPOND_80100313  \\\n",
       "gene                        ...                                       \n",
       "ARHGEF16          4.700440  ...          6.768184          5.754888   \n",
       "HMG20B            8.299208  ...         10.392317         10.890264   \n",
       "PLXNB2            9.169925  ...          9.611025         10.292322   \n",
       "WFIKKN1           0.857649  ...          3.906891          3.321928   \n",
       "MC1R              4.392317  ...          6.629357          6.523562   \n",
       "...                    ...  ...               ...               ...   \n",
       "MBD3              6.247928  ...          7.033423          7.434628   \n",
       "CYB561D1          2.584963  ...          2.584963          4.392317   \n",
       "GUCY1B1           3.584963  ...          2.584963          1.584963   \n",
       "KLF1              0.000000  ...          0.102749          1.000000   \n",
       "MEGF6             5.209453  ...          6.894818          5.000000   \n",
       "\n",
       "          RESPOND_80100345  RESPOND_80100411  RESPOND_80100526  \\\n",
       "gene                                                             \n",
       "ARHGEF16          3.459432          7.179909          6.794416   \n",
       "HMG20B            8.499846         11.106563          9.874981   \n",
       "PLXNB2            8.810572         10.650154         10.079485   \n",
       "WFIKKN1           1.000000          5.426265          3.459432   \n",
       "MC1R              4.584963          7.965784          7.483816   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              5.169925          8.370687          8.118941   \n",
       "CYB561D1          2.807355          4.392317          4.321928   \n",
       "GUCY1B1           4.392317          2.000000          2.584963   \n",
       "KLF1              0.000000          0.852043          0.879822   \n",
       "MEGF6             6.087463          5.247928          5.129283   \n",
       "\n",
       "          RESPOND_80100556  RESPOND_80100590  RESPOND_81100031  \\\n",
       "gene                                                             \n",
       "ARHGEF16          3.700440          5.169925          6.285402   \n",
       "HMG20B            9.154818          9.491853          9.292322   \n",
       "PLXNB2            8.625709          9.337622          9.612868   \n",
       "WFIKKN1           0.000000          2.000000          3.584963   \n",
       "MC1R              2.321928          7.523562          6.392317   \n",
       "...                    ...               ...               ...   \n",
       "MBD3              4.954196          7.011227          7.977280   \n",
       "CYB561D1          4.087463          3.807355          3.459432   \n",
       "GUCY1B1           3.584963          3.169925          3.000000   \n",
       "KLF1              0.000000          0.042794          1.000000   \n",
       "MEGF6             5.000000          5.087463          7.022368   \n",
       "\n",
       "          RESPOND_40100842  RESPOND_80100259  \n",
       "gene                                          \n",
       "ARHGEF16          5.807355          6.303781  \n",
       "HMG20B            9.296916         10.213104  \n",
       "PLXNB2            9.317413          9.569856  \n",
       "WFIKKN1           3.459432          3.459432  \n",
       "MC1R              6.209453          7.475733  \n",
       "...                    ...               ...  \n",
       "MBD3              6.700440          7.033423  \n",
       "CYB561D1          3.906891          3.906891  \n",
       "GUCY1B1           3.807355          2.807355  \n",
       "KLF1              0.000000          0.078879  \n",
       "MEGF6             6.022368          5.285402  \n",
       "\n",
       "[100 rows x 164 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "heatmap_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "74d85b53-9bb0-478f-a08d-d7d0f8186c41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row Silhouette Score: 0.24124544217953459\n",
      "Column Silhouette Score: 0.06585128389732592\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if not hasattr(array, \"sparse\") and array.dtypes.apply(is_sparse).any():\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if not hasattr(array, \"sparse\") and array.dtypes.apply(is_sparse).any():\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.\n",
      "  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):\n"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.cluster.hierarchy import fcluster, linkage\n",
    "from sklearn.metrics import silhouette_score\n",
    "\n",
    "def hierarchical_clustering(expression_df_heatmap, row_threshold=7, col_threshold=7, output_folder=target_gene):\n",
    "    # Cluster the rows and columns using hierarchical clustering\n",
    "    row_linkage = linkage(expression_df_heatmap, method='ward')\n",
    "    col_linkage = linkage(expression_df_heatmap.T, method='ward')\n",
    "\n",
    "    # Assign cluster labels using cluster\n",
    "    row_clusters = fcluster(row_linkage, t=row_threshold, criterion='maxclust')\n",
    "    col_clusters = fcluster(col_linkage, t=col_threshold, criterion='maxclust')\n",
    "\n",
    "    # Calculate silhouette scores\n",
    "    row_silhouette_score = silhouette_score(expression_df_heatmap, row_clusters)\n",
    "    col_silhouette_score = silhouette_score(expression_df_heatmap.T, col_clusters)\n",
    "\n",
    "    print(f\"Row Silhouette Score: {row_silhouette_score}\")\n",
    "    print(f\"Column Silhouette Score: {col_silhouette_score}\")\n",
    "\n",
    "    # Create dictionaries to store cluster information\n",
    "    row_cluster_info = {f\"Cluster {cluster}\": expression_df_heatmap.index[row_clusters == cluster] for cluster in np.unique(row_clusters)}\n",
    "    col_cluster_info = {f\"Cluster {cluster}\": expression_df_heatmap.columns[col_clusters == cluster] for cluster in np.unique(col_clusters)}\n",
    "\n",
    "    row_cluster_file = os.path.join(output_folder, 'row_clusters.txt')\n",
    "    with open(row_cluster_file, 'w') as file:\n",
    "        file.write(f\"Row Silhouette Score: {row_silhouette_score}\\n\\n\")\n",
    "        for cluster, items in row_cluster_info.items():\n",
    "            file.write(f\"{cluster}:\\n\")\n",
    "            file.write(f\"{', '.join(items)}\\n\\n\")\n",
    "\n",
    "    # Save column cluster information to a text file\n",
    "    col_cluster_file = os.path.join(output_folder, 'col_clusters.txt')\n",
    "    with open(col_cluster_file, 'w') as file:\n",
    "        file.write(f\"Column Silhouette Score: {col_silhouette_score}\\n\\n\")\n",
    "        for cluster, items in col_cluster_info.items():\n",
    "            file.write(f\"{cluster}:\\n\")\n",
    "            file.write(f\"{', '.join(items)}\\n\\n\")\n",
    "    \n",
    "    return row_linkage, col_linkage, row_cluster_info, col_cluster_info\n",
    "\n",
    "\n",
    "row_linkage, col_linkage, row_clusters, col_clusters = hierarchical_clustering(heatmap_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "07391585-1787-4d24-85c9-8c597915f80c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def create_clustered_heatmap_and_save(expression_df_heatmap, \n",
    "                                      row_linkage, \n",
    "                                      col_linkage, \n",
    "                                      sample_mutation_df, \n",
    "                                      output_folder=target_gene):\n",
    "    # Group labels and colors for color bar\n",
    "    type_map = {1: 'red', 0: 'yellow'}\n",
    "\n",
    "    sns.set(font_scale=1.0) \n",
    "\n",
    "    # Create a clustered heatmap\n",
    "    clustered_df = sns.clustermap(\n",
    "        expression_df_heatmap,\n",
    "        row_linkage=row_linkage,\n",
    "        col_linkage=col_linkage,\n",
    "        cmap='viridis',\n",
    "        annot=False,\n",
    "        fmt=\".1f\",\n",
    "        linewidths=.5,\n",
    "        col_colors=sample_mutation_df.set_index('Sample')['Mutation Status'].map(type_map),\n",
    "        z_score=0,\n",
    "        cbar_kws={\"shrink\": 0.7, \"aspect\": 30}  # Adjust color bar size\n",
    "    )\n",
    "\n",
    "    #Set the size of the overall figure\n",
    "    clustered_df.fig.set_size_inches(15, len(expression_df_heatmap) * 0.2)  \n",
    "\n",
    "    output_filename = f'{output_folder}/heatmap.png'\n",
    "\n",
    "    clustered_df.ax_heatmap.set_yticklabels(clustered_df.ax_heatmap.get_yticklabels(), rotation=0)\n",
    "    \n",
    "    # Save the plot to a file\n",
    "    clustered_df.savefig(output_filename, bbox_inches='tight')\n",
    "\n",
    "    # Close the plot to prevent displaying it in the notebook (optional)\n",
    "    plt.close()\n",
    "\n",
    "    return clustered_df\n",
    "\n",
    "clustered_heatmap = create_clustered_heatmap_and_save(heatmap_data, \n",
    "                                                      row_linkage, \n",
    "                                                      col_linkage,\n",
    "                                                      mutated_status_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "56c644ef-b34e-4343-b27e-e281a2be9aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram\n",
    "\n",
    "## add more args to the function \n",
    "\n",
    "def plot_and_save_dendrograms(row_linkage, col_linkage, expression_df_heatmap, output_folder=target_gene):\n",
    "    # Plot the row dendrogram\n",
    "    plt.figure(figsize=(12, 10))\n",
    "    # play around with fig sizes \n",
    "    gene_plot = dendrogram(row_linkage, \n",
    "                           labels=expression_df_heatmap.index, \n",
    "                           truncate_mode='lastp', \n",
    "                           p=int(0.75 * len(row_linkage)), show_leaf_counts=True)\n",
    "    plt.title('Row Dendrogram')\n",
    "    \n",
    "    # Save the row dendrogram plot to a file\n",
    "    row_dendrogram_filename = f'{output_folder}/row_dendrogram.png'\n",
    "    plt.savefig(row_dendrogram_filename)\n",
    "    plt.close()\n",
    "\n",
    "    # Plot the column dendrogram\n",
    "    # play around with fig sizes \n",
    "    plt.figure(figsize=(12, 10))\n",
    "    sample_plot = dendrogram(col_linkage, labels=expression_df_heatmap.columns, orientation='top', \n",
    "                             truncate_mode='lastp', p=int(0.75 * len(row_linkage)), show_leaf_counts=True)\n",
    "    plt.title('Column Dendrogram')\n",
    "    \n",
    "    # Save the column dendrogram plot to a file\n",
    "    col_dendrogram_filename = f'{output_folder}/col_dendrogram.png'\n",
    "    plt.savefig(col_dendrogram_filename)\n",
    "    plt.close()\n",
    "\n",
    "    return gene_plot, sample_plot\n",
    "\n",
    "row_dendrogram, col_dendrogram = plot_and_save_dendrograms(row_linkage, \n",
    "                                                           col_linkage, \n",
    "                                                           heatmap_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "0e7b9c4a-c204-40a4-af6f-a80cbb4a0594",
   "metadata": {},
   "outputs": [],
   "source": [
    "def histogram_of_column_and_save(df, column, output_folder=target_gene):\n",
    "    # Plot the distribution of p-values\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.hist(df[column], bins=30, color='blue', edgecolor='black')\n",
    "    plt.title(f'Distribution of {column}')\n",
    "    plt.xlabel('P-values')\n",
    "    plt.ylabel('Frequency')\n",
    "\n",
    "    output_filename = f'{output_folder}/histogram_{column}.png'\n",
    "    # Save the histogram plot to a file\n",
    "    plt.savefig(output_filename)\n",
    "    plt.close()\n",
    "\n",
    "histogram_of_column_and_save(volcano_plot_df, 'pvalue')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "0088ccb1-ed8b-4faa-a98f-4f69963bc1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def volcano_plot(input_file_path, yaxis, \n",
    "                 xaxis, \n",
    "                 significance_threshold=0.05, \n",
    "                 logfold_positive_threshold=2, \n",
    "                 logfold_negative_threshold=-2,\n",
    "                 output_folder=target_gene):\n",
    "    df = pd.read_csv(input_file_path)\n",
    "\n",
    "    # Apply -log10 transformation to the p-value\n",
    "    df['-log10_pvalue'] = -np.log10(df[yaxis])\n",
    "\n",
    "    # Highlight significant points with large log-fold changes\n",
    "    significant_genes_positive = df[(df[yaxis] < significance_threshold) & (df['logFC'] > logfold_positive_threshold)]\n",
    "    significant_genes_negative = df[(df[yaxis] < significance_threshold) & (df['logFC'] < logfold_negative_threshold)]\n",
    "\n",
    "    # Clip outliers in 'logFC' column for the plot\n",
    "    df['logFC'] = df['logFC'].clip(lower=-10, upper=10)\n",
    "\n",
    "    plt.scatter(x=df[xaxis], y=df['-log10_pvalue'], s=1, label='All Genes', alpha=0.5)\n",
    "    plt.scatter(x=significant_genes_positive[xaxis], y=significant_genes_positive['-log10_pvalue'], s=10, c='red', marker='^', label='Significant Genes (Positive LogFC)')\n",
    "    plt.scatter(x=significant_genes_negative[xaxis], y=significant_genes_negative['-log10_pvalue'], s=10, c='blue', marker='v', label='Significant Genes (Negative LogFC)')\n",
    "\n",
    "    plt.xlabel(xaxis)\n",
    "    plt.ylabel(f'-log10({yaxis})')\n",
    "    plt.title('Volcano Plot')\n",
    "    plt.axhline(-np.log10(significance_threshold), color='gray', linestyle='--', label=f'Significance Threshold ({significance_threshold})')\n",
    "    plt.axvline(logfold_positive_threshold, color='gray', linestyle='--', label=f'Positive Log-Fold Change Threshold ({logfold_positive_threshold})')\n",
    "    plt.axvline(logfold_negative_threshold, color='gray', linestyle='--', label=f'Negative Log-Fold Change Threshold ({logfold_negative_threshold})')\n",
    "    \n",
    "     # Create a separate legend outside the plot\n",
    "    fig = plt.gcf()\n",
    "    handles, labels = plt.gca().get_legend_handles_labels()\n",
    "    fig.legend(handles, labels, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    \n",
    "    \n",
    "    output_filename = f'{output_folder}/volcano_plot.png'\n",
    "    plt.savefig(output_filename, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    significant_genes_positive.to_csv(f'{output_folder}/signif_genes_positive.csv')\n",
    "    significant_genes_negative.to_csv(f'{output_folder}/signif_genes_negative.csv')\n",
    "\n",
    "    return significant_genes_positive, significant_genes_negative\n",
    "\n",
    "significant_genes_positive, significant_genes_negative  = volcano_plot(volcano_input_filename, 'pvalue', 'logFC')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "7f2a377a-206b-4256-a9ed-39d78bfe3bf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/Users/meltemtutar/miniconda3/envs/huang_3.9/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def create_gene_expression_boxplot(expression_df, significant_genes_df, mutated_status_df, output_folder=target_gene, positive=1 ):\n",
    "    # Filter for significant genes\n",
    "    filtered_expression_df_genes_of_interest = expression_df.loc[significant_genes_df.gene]\n",
    "\n",
    "    # Unstack the expression DataFrame\n",
    "    unstack_expression = filtered_expression_df_genes_of_interest.unstack().reset_index()\n",
    "    unstack_expression.columns = ['Sample', 'Gene', 'Expression']\n",
    "\n",
    "    # Merge the expression matrix with the mutation information\n",
    "    boxplot_df = pd.merge(unstack_expression, mutated_status_df, how='inner', on='Sample')\n",
    "\n",
    "    # Rename columns for clarity\n",
    "    boxplot_df.columns = ['Sample', 'Gene', 'Expression', 'Target Gene Mutation Status']\n",
    "\n",
    "    # Set up the plot\n",
    "    plt.figure(figsize=(12, 8))\n",
    "\n",
    "    # Create a box plot with 'hue' for each gene and 'dodge' for Mutation Status\n",
    "    sns.boxplot(x='Gene', y='Expression', data=boxplot_df, hue='Target Gene Mutation Status', dodge=True)\n",
    "\n",
    "    # Customize the plot\n",
    "    plt.xlabel('Gene')\n",
    "    plt.ylabel('Gene Expression')\n",
    "    plt.title('Box Plots of Gene Expression for Mutated and Non-Mutated Individuals')\n",
    "\n",
    "    # Rotate x-axis labels for better readability\n",
    "    plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "    if positive == 1:\n",
    "        output_filename = f'{output_folder}/positive_genes_expression_boxplot.png'\n",
    "    else:\n",
    "        output_filename = f'{output_folder}/negative_genes_expression_boxplot.png'\n",
    "    plt.savefig(output_filename)\n",
    "    plt.close()\n",
    "\n",
    "create_gene_expression_boxplot(expression_df, significant_genes_positive, mutated_status_df, positive=1)\n",
    "create_gene_expression_boxplot(expression_df, significant_genes_negative, mutated_status_df, positive=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "765c7969-47bd-4c9f-8055-958e5fb1ca6a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
